<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.2.313">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Marguerite Butler">
<meta name="dcterms.date" content="2023-04-04">
<meta name="description" content="Getting started with morphometrics">

<title>ZOOL710 Data Science in R for Biologists - Intro to Geometric Morphometrics</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>

  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<link rel="stylesheet" href="../../styles.css">
</head>

<body class="floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">ZOOL710 Data Science in R for Biologists</span>
    </a>
  </div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../index.html">
 <span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-general-information" role="button" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">General Information</span>
    </a>
    <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="nav-menu-general-information">    
        <li>
    <a class="dropdown-item" href="../../syllabus.html">
 <span class="dropdown-text">Syllabus</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../schedule.html">
 <span class="dropdown-text">Schedule</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-course-materials" role="button" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Course Materials</span>
    </a>
    <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="nav-menu-course-materials">    
        <li>
    <a class="dropdown-item" href="../../lectures.html">
 <span class="dropdown-text">Lectures</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../projects.html">
 <span class="dropdown-text">Projects</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../resources.html">
 <span class="dropdown-text">Resources</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/"><i class="bi bi-github" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://twitter.com"><i class="bi bi-twitter" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="../../index.xml"><i class="bi bi-rss" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
</ul>
              <div id="quarto-search" class="" title="Search"></div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<header id="title-block-header" class="quarto-title-block default toc-left page-columns page-full">
  <div class="quarto-title-banner page-columns page-full">
    <div class="quarto-title column-body">
      <h1 class="title">Intro to Geometric Morphometrics</h1>
                  <div>
        <div class="description">
          Getting started with morphometrics
        </div>
      </div>
                          <div class="quarto-categories">
                <div class="quarto-category">module 6</div>
                <div class="quarto-category">week 12</div>
                <div class="quarto-category">control structures</div>
                <div class="quarto-category">if else</div>
                <div class="quarto-category">(do) while</div>
                <div class="quarto-category">programming</div>
              </div>
                  </div>
  </div>
    
  <div class="quarto-title-meta-author">
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-heading">Affiliation</div>
    
      <div class="quarto-title-meta-contents">
      <a href="https://butlerlab.org">Marguerite Butler</a> 
    </div>
      <div class="quarto-title-meta-contents">
          <p class="affiliation">
              School of Life Sciences, University of Hawaii
            </p>
        </div>
      </div>

  <div class="quarto-title-meta">

        
      <div>
      <div class="quarto-title-meta-heading">Published</div>
      <div class="quarto-title-meta-contents">
        <p class="date">April 4, 2023</p>
      </div>
    </div>
    
      
    </div>
    
  
  </header><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse sidebar-navigation floating overflow-auto">
    <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#acknowledgements" id="toc-acknowledgements" class="nav-link active" data-scroll-target="#acknowledgements">Acknowledgements</a></li>
  <li><a href="#learning-objectives" id="toc-learning-objectives" class="nav-link" data-scroll-target="#learning-objectives">Learning objectives</a>
  <ul class="collapse">
  <li><a href="#preparation" id="toc-preparation" class="nav-link" data-scroll-target="#preparation">Preparation</a></li>
  </ul></li>
  <li><a href="#overview" id="toc-overview" class="nav-link" data-scroll-target="#overview">Overview</a></li>
  <li><a href="#capturing-images" id="toc-capturing-images" class="nav-link" data-scroll-target="#capturing-images">Capturing images</a></li>
  <li><a href="#landmarks" id="toc-landmarks" class="nav-link" data-scroll-target="#landmarks">Landmarks</a>
  <ul class="collapse">
  <li><a href="#digitizing-landmarks-on-images" id="toc-digitizing-landmarks-on-images" class="nav-link" data-scroll-target="#digitizing-landmarks-on-images">Digitizing Landmarks on Images</a></li>
  </ul></li>
  <li><a href="#geomorph-package" id="toc-geomorph-package" class="nav-link" data-scroll-target="#geomorph-package"><code>geomorph</code> package</a>
  <ul class="collapse">
  <li><a href="#digitizing-landmarks-in-geomorph" id="toc-digitizing-landmarks-in-geomorph" class="nav-link" data-scroll-target="#digitizing-landmarks-in-geomorph">Digitizing landmarks in <code>geomorph</code></a></li>
  </ul></li>
  <li><a href="#tps-format" id="toc-tps-format" class="nav-link" data-scroll-target="#tps-format">TPS format</a></li>
  <li><a href="#semi-landmarks" id="toc-semi-landmarks" class="nav-link" data-scroll-target="#semi-landmarks">Semi-Landmarks</a></li>
  <li><a href="#size-adjustment-with-procrustes-superimposition" id="toc-size-adjustment-with-procrustes-superimposition" class="nav-link" data-scroll-target="#size-adjustment-with-procrustes-superimposition">Size Adjustment with Procrustes Superimposition</a>
  <ul class="collapse">
  <li><a href="#centroid" id="toc-centroid" class="nav-link" data-scroll-target="#centroid">Centroid</a></li>
  <li><a href="#centroid-size" id="toc-centroid-size" class="nav-link" data-scroll-target="#centroid-size">Centroid size</a></li>
  <li><a href="#shape-variables" id="toc-shape-variables" class="nav-link" data-scroll-target="#shape-variables">Shape Variables</a></li>
  </ul></li>
  <li><a href="#geomorph-package-1" id="toc-geomorph-package-1" class="nav-link" data-scroll-target="#geomorph-package-1">Geomorph Package</a>
  <ul class="collapse">
  <li><a href="#data-structures" id="toc-data-structures" class="nav-link" data-scroll-target="#data-structures">Data Structures</a></li>
  <li><a href="#built-in-example" id="toc-built-in-example" class="nav-link" data-scroll-target="#built-in-example">Built-in Example</a></li>
  </ul></li>
  <li><a href="#more-software" id="toc-more-software" class="nav-link" data-scroll-target="#more-software">More software</a></li>
  </ul>
</nav>
</nav>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
    </div>
<!-- main -->
<main class="content quarto-banner-title-block" id="quarto-document-content">




<section id="acknowledgements" class="level3">
<h3 class="anchored" data-anchor-id="acknowledgements">Acknowledgements</h3>
<p>References for this Material:</p>
<ul>
<li>Bardua et al (2019) A Practical Guide to Sliding and Surface Semilandmarks in Morphometric Analyses<a href="https://academic.oup.com/iob/article/1/1/obz016/5526881" class="uri">https://academic.oup.com/iob/article/1/1/obz016/5526881</a></li>
<li><a href="https://www.researchgate.net/profile/Alexandre-Palaoro/post/Geomorph-Package-in-R/attachment/5cffb777cfe4a7968da98557/AS%3A768610867900420%401560262519604/download/Quick_Guide_to_Geomorph-3.0.pdf">Emma Sherratʻs quick guide to Geomorph</a></li>
</ul>
</section>
<section id="learning-objectives" class="level1">
<h1>Learning objectives</h1>
<div class="callout-note callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Learning objectives
</div>
</div>
<div class="callout-body-container callout-body">
<p><strong>At the end of this lesson you will:</strong></p>
<ul>
<li>Be able to collect landmark data</li>
<li>Understand the terminology for landmark data</li>
</ul>
</div>
</div>
<section id="preparation" class="level3">
<h3 class="anchored" data-anchor-id="preparation">Preparation</h3>
<p>Packages:<br>
<strong>geomorph</strong><br>
<strong>rgl</strong><br>
<strong>Matrix</strong><br>
Please install if you donʻt have them. <code>install.packages(c("geomorph","rgl","Matrix"), dependencies=T)</code></p>
<p>Mac Users: Please also install Xquartz <a href="https://www.xquartz.org" class="uri">https://www.xquartz.org</a>, a windowing system for <code>rgl</code>.</p>
<p>Files:<br>
<code>red-fox-mandible-lateral.jpg</code> Please <a href="../../images/red-fox-mandible-lateral.jpg">download</a> into your folder for this exercise. <a href="https://russellboneatlas.wordpress.com/home/red-fox-bone-atlas/red-fox-mandible/">Source: Jim Russell</a></p>
</section>
</section>
<section id="overview" class="level1">
<h1>Overview</h1>
<p><strong>Traditional morphometrics</strong> uses linear measurements such as body length, hind limb length, etc. to characterize morphological variation. This works well for many questions, especially ones that relate to lever mechanics, such as locomotion or bite force, etc. but does not capture 3D variation well such as the geometric shape of a skull or the details of seed shape.</p>
<p><strong>Geometric morphometrics</strong> aims to characterize more complex shapes by the use of <strong>landmarks</strong>, locations on the specimens determined by the researcher to reveal key insights or capture the full range of variation.</p>
<p>In order to do this, we still need methods to characterize <strong>size</strong> vs.&nbsp;<strong>shape</strong>. For geometric morphometrics that is typically <strong>Procrustes superposition</strong>. Some other differences also involve the data. Whereas one can typically take linear measurements directly on the specimen (with calipers or a ruler), collecting landmark data usually involves some image capture.</p>
</section>
<section id="capturing-images" class="level1">
<h1>Capturing images</h1>
<p>The most important thing is consistency and a protocol that minimizes error. While the specific protocol will vary by the type of object and technology (cameras, uCT, etc.), some general principles are :</p>
<ul>
<li>Specimen Layout
<ul>
<li>Be as <strong>consistent</strong> as possible in the preparation of specimens and their arrangement</li>
<li>Place every specimen on the same side. (Capture the same aspect of each specimen)</li>
<li>Same pose or posture</li>
<li>Mouths open? Closed?</li>
<li>Fins or dewlaps exposed to show shape and articulatio, etc.</li>
</ul></li>
<li>Size and Metadata
<ul>
<li>Include a size standard (e.g., a ruler), ideally the same item, with X-Y information</li>
<li>Include the specimen ID in the photo itself (written on a card, etc.)</li>
</ul></li>
<li>Camera
<ul>
<li>Camera in same position and same angle, preferably with a tripod</li>
<li>Use the same camera and lens (or zoom, etc.)</li>
<li>Make several backups (multiple flash cards, save to computer, etc.)</li>
<li>Specimen perpendicular to camera (check level)</li>
<li>Same lighting</li>
<li>Take several photos as backup</li>
</ul></li>
<li>Data Management
<ul>
<li>Keep a notebook (written log) of the specimens being photographed, in order</li>
<li>Transfer files to raw data folder for photos (in same format and same order)</li>
<li>Name files with metadata in filenames for each specimen, save in another folder</li>
<li>You should have Raw_Photos, Named_Photos folders.</li>
</ul></li>
</ul>
<p>Errors and distortions can be fixed with software, but it is much easier to minimize them at the start.</p>
</section>
<section id="landmarks" class="level1">
<h1>Landmarks</h1>
<p>Essential features of good landmarks:</p>
<p>Choose landmarks that are:</p>
<ul>
<li>Present on all specimens<br>
</li>
<li>Relevant to the question<br>
</li>
<li>Clearly defined (to be repeatable)<br>
</li>
<li>Conform to accepted best practices for your type of study (do a literature search)</li>
</ul>
<p>Here is an example of typical landmarks:</p>
<p><img src="../../images/rodentmandible.jpg" class="img-fluid" alt="Rodent lower jaw (mandible). Lateral view of a Nectomys squamipes mandible, showing the 10 landmarks used in this work. 1) Anterior edge of the incisive alveolus; 2) Most posterior point of the diastema; 3) Junction between the mandible and the m1 root; 4) Junction between the molar base and the coronoid process; 5) Most posterior point at the articular process; 6) Most anterior point in the angular notch; 7) Most posterior point at the angular process; 8) Angular process base; 9) Inner edge of the mandibular symphysis; 10) Posterior edge of the incisive alveolus"> <a href="https://onlinelibrary.wiley.com/doi/full/10.1002/ece3.7557">Source</a></p>
<section id="digitizing-landmarks-on-images" class="level3">
<h3 class="anchored" data-anchor-id="digitizing-landmarks-on-images">Digitizing Landmarks on Images</h3>
<p>Acquiring landmarks usually involves some digitizing by hand (with computer assistance). For 2D landmark data the workflow is:</p>
<ul>
<li>Read in the image file<br>
</li>
<li>Set the scale<br>
</li>
<li>Digitize landmarks, collecting X-Y coordinate data</li>
</ul>
<p>Image files are rows and columns of pixels, the number of which is determined by the file size and resolution. Pixels therefore, form a grid which we can digitize, meaning we can indicate at which pixel our landmark of interest is. We translate the pixel coordinates into real-world distances, typically by ditigizing an object of known dimensions such as a ruler. Digitizing software typically asks the user to indicate the points on the screen and the known distances they correspond to.</p>
<p>Easy-to use software for this purpose include <code>ImageJ</code> and <code>FIJI</code>, as well as the <code>geomorph</code> package in R. Before doing a large project, I would try them all to see which serves your purpose the best.</p>
</section>
</section>
<section id="geomorph-package" class="level1">
<h1><code>geomorph</code> package</h1>
<p>The package <code>geomorph</code> is widely used for geometric morphometrics, and will serve as a good example to illustrate the geometric morphometrics workflow all in R (depending on the project and the material, however, you may find it more convenient or efficient to use other software for digitizing).</p>
<section id="digitizing-landmarks-in-geomorph" class="level3">
<h3 class="anchored" data-anchor-id="digitizing-landmarks-in-geomorph">Digitizing landmarks in <code>geomorph</code></h3>
<p>For a quick example, letʻs do a 2D landmark collection on the Red Fox Mandible using <code>digitize2d()</code>. Emma Sherratʻs <a href="https://www.researchgate.net/profile/Alexandre-Palaoro/post/Geomorph-Package-in-R/attachment/5cffb777cfe4a7968da98557/AS%3A768610867900420%401560262519604/download/Quick_Guide_to_Geomorph-3.0.pdf">quick guide to Geomorph</a> provides a more thorough walk-through for capturing landmarks in Chapter 15.</p>
<p>The syntax for <code>digitize2d()</code> is:</p>
<p><code>digitize2d(filelist, nlandmarks, scale = NULL, tpsfile, verbose = TRUE)</code></p>
<p>Where:</p>
<ul>
<li><code>filelist</code> : a vector of filenames of image files</li>
<li><code>nlandmarks</code> : the number of landmarks to digitize<br>
</li>
<li><code>scale</code> : the length of scale to use (optional)<br>
</li>
<li><code>tpsfile</code> : the name of an output file to create or read</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">require</span>(geomorph)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>files <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"red-fox-mandible-lateral.jpg"</span>)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="co"># files &lt;- list.files(pattern = "*.jpg") </span></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>             <span class="co"># for multiple files in your image directory</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">digitize2d</span>( files, </span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>            <span class="at">nlandmarks=</span><span class="dv">10</span>, </span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>            <span class="at">scale=</span><span class="dv">80</span>, </span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>            <span class="at">tpsfile =</span> <span class="st">"mandible-lat.tps"</span>, </span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>            <span class="at">verbose =</span> <span class="cn">TRUE</span>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>R will now wait for you to digitize the first landmark. In verbose mode, it will ask you to <em>confirm</em> each landmark (this can be turned off). If a landmark is missing, choose <code>a</code> for absent. If you make a mistake and choose <code>n</code>, it will ask you to redigitize it. The output will be sent to a <code>.tps</code>file.</p>
</section>
</section>
<section id="tps-format" class="level1">
<h1>TPS format</h1>
<p>TPS format is a standard morphometric data format for storing landmark coordinate data.</p>
<p><code>readland.tps()</code> reads TPS format and returns an array: <em>p</em> landmarks by <em>k</em> dimensions by <em>N</em> specimens.</p>
<p>Remember an array is simply a multidimensional dataframe. Think of it here as a stack of spreadsheets. Each spreadsheet has rows for landmarks and columns for dimensions. We have a separate spreadsheet for each specimen, making this a 3-dimensional array.</p>
<p>The third dimension of this array contains names for each specimen, which are obtained from the image names in the <em>.tps</em> file. We can read in the data that we digitized above:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>dat <span class="ot">&lt;-</span> <span class="fu">readland.tps</span>(<span class="st">"mandible-lat.tps"</span>, <span class="at">specID =</span> <span class="st">"ID"</span>, <span class="at">readcurves =</span> <span class="cn">FALSE</span>, <span class="at">warnmsg =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
No curves detected; all points appear to be fixed landmarks.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>dat</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>, , ../../images/red-fox-mandible-lateral.jpg

           [,1]      [,2]
 [1,] 107.29125 109.47242
 [2,] 103.69036 108.94577
 [3,] 101.66872 108.37825
 [4,]  43.63469 105.62467
 [5,]  10.59532 111.73118
 [6,]  16.56440 102.17078
 [7,]  12.09659  98.37737
 [8,]  34.05770  93.06904
 [9,]  66.20128  89.99578
[10,] 113.50334 107.91870</code></pre>
</div>
</div>
<p>Note that the name saved with the data in this case was the file name. For another example, check out the <code>plethodon</code> data included with <code>geomorph</code></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(plethodon)</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="fu">attributes</span>(plethodon)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$names
[1] "land"    "links"   "species" "site"    "outline"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(plethodon)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>List of 5
 $ land   : num [1:12, 1:2, 1:40] 8.89 9.27 5.56 1.87 1.28 ...
  ..- attr(*, "dimnames")=List of 3
  .. ..$ : NULL
  .. ..$ : NULL
  .. ..$ : NULL
 $ links  : num [1:14, 1:2] 4 3 2 1 1 6 7 8 9 10 ...
 $ species: Factor w/ 2 levels "Jord","Teyah": 1 1 1 1 1 1 1 1 1 1 ...
 $ site   : Factor w/ 2 levels "Allo","Symp": 2 2 2 2 2 2 2 2 2 2 ...
 $ outline: num [1:3631, 1:2] 0.399 0.4 0.401 0.403 0.404 ...</code></pre>
</div>
</div>
<p>3D coordinates in tps format simply have a third dimension (z in addition to x and y) in each spreadsheet (one spreadsheet or table per specimen).</p>
<p><img src="../../images/geomorphdataarrays.png" class="img-fluid"></p>
<p><a href="https://www.researchgate.net/profile/Alexandre-Palaoro/post/Geomorph-Package-in-R/attachment/5cffb777cfe4a7968da98557/AS%3A768610867900420%401560262519604/download/Quick_Guide_to_Geomorph-3.0.pdf">Source: Emma Sherratʻs quick guide to Geomorph</a></p>
</section>
<section id="semi-landmarks" class="level1">
<h1>Semi-Landmarks</h1>
<p>Oftentimes, there is more shape variation than what is captured by the landmarks themselves, which are only points. In addition, sometimes curves and surfaces have no distinct feature that can be reliably captured as the same homologous point from specimen to specimen. For example, the curvature of a smooth shell may lack any distinct landmarks. In both of these cases, semi-landmarks can be helpful.</p>
<p><strong>Semi-landmarks</strong> are regularly spaced points between two landmarks or along a curve or a surface.<br>
<strong>Sliding semi-landmarks</strong> are spaced linearly along a curve.<br>
<strong>Surface semi-landmarks</strong> are spread out evenly in 3D along a surface.</p>
<p>NOTE: One thing to note is that any error in the landmarks will also be propagated to the semilandmarks.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="../../images/caecelian_skull_semilandmarks.jpeg" class="img-fluid figure-img"></p>
<p></p><figcaption class="figure-caption">Figure: Landmark and semilandmark data displayed on the caecilian <em>Siphonops annulatus</em> BMNH 1956.1.15.88. Points are colored as follows: landmarks (red), sliding semilandmarks (“curve points,” yellow), and surface semilandmarks (“surface points,” blue). For information regarding each cranial region, see Bardua et al.&nbsp;(2019). BMNH, Natural History Museum, London, UK.</figcaption><p></p>
</figure>
</div>
<p><a href="https://academic.oup.com/iob/article/1/1/obz016/5526881">Source: Bardua et al 2019</a></p>
<p>Annotated 3D version of this figure available at: <a href="https://sketchfab.com/3d-models/add35e2e8af94839b1f577bfcee32e54" class="uri">https://sketchfab.com/3d-models/add35e2e8af94839b1f577bfcee32e54</a></p>
<p>Letʻs practice digitizing 3D landmarks with the geomorph vignette: <a href="https://cran.r-project.org/web/packages/geomorph/vignettes/geomorph.digitize3D.html" class="uri">https://cran.r-project.org/web/packages/geomorph/vignettes/geomorph.digitize3D.html</a></p>
</section>
<section id="size-adjustment-with-procrustes-superimposition" class="level1">
<h1>Size Adjustment with Procrustes Superimposition</h1>
<p>Procrustes superimposition is the most widely used method to create size and shape variables from landmark data. The basic idea is that larger objects should have landmarks that are farther apart than the same landmarks in smaller objects.</p>
<section id="centroid" class="level3">
<h3 class="anchored" data-anchor-id="centroid">Centroid</h3>
<p>A key concept is the <strong>centroid</strong>, which the center point of all of the landmarks. It is easily calculated for each specimen by taking the average of all of the landmark coordinates along each axis. For <em>p</em> landmarks in two dimensions the coordinates of the centroid would be:</p>
<p><span class="math display">\[
C_{x} = \frac{X_1+X_2+X_3+ ... + X_p}{p}
\]</span> <span class="math display">\[
C_{y} = \frac{Y_1+Y_2+Y_3 ... + Y_p}{p}
\]</span></p>
</section>
<section id="centroid-size" class="level3">
<h3 class="anchored" data-anchor-id="centroid-size">Centroid size</h3>
<p>The concept of <strong>centroid size</strong> describes geometric size. It is calculated from the distances of each landmark from the centroid, but it is a geometric mean because it is caculated by taking the square-root of the sums of squared distances of each landmark from the centroid.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="../../images/Triangle.Centroid.png" class="img-fluid figure-img"></p>
<p></p><figcaption class="figure-caption">The centroid of a triangle (with landmarks at the vertices) is the point in the center.</figcaption><p></p>
</figure>
</div>
<p>[Source: Wikipedia]</p>
</section>
<section id="shape-variables" class="level3">
<h3 class="anchored" data-anchor-id="shape-variables">Shape Variables</h3>
<p>Shape variables calculated from Procrustes superimposition mathematically separate scale (size) from shape by applying shape-preserving transformations to make the landmarks as similar as possible.</p>
<p>The concept: In order to compare differences in shape, we can remove size (in other words, normalize by size or bring all of the specimens to the same size) with the following transformations:</p>
<ul>
<li><strong>Scale</strong>: Bring each set of landmarks to the same size by magnifing or reducing the distances of the landmarks around their centroid.<br>
</li>
<li><strong>Translate</strong>: Shift the landmarks in the coordinate space to place the centroids at a common point.<br>
</li>
<li><strong>Rotate</strong>: Rotate the landmarks with an angular transformation to bring them into a common orientation.</li>
</ul>
<p>At the end of Procrustes superimposition, the landmarks for the different specimens will be scaled to the same size and their coordinates will lie as close together as possible while preserving the differences in shape. The figure below illustrates these steps.</p>
<p>The mathematics of these transofrmations come from basic linear algebra operations on matrices, which are beyond the scope of the class, but to understand what is happening we can analogize it to univariate arithmetic.</p>
<p><em>Scaling</em> is like dividing each variable by a size variable to put all of the individuals on the same baseline with regard to size (magnitude).</p>
<p><em>Translation</em> brings the centers of all of the landmarks into alignment. Think of a situation where we measured the length of specimens according to wherever they fell within the picture frame, for example if one specimen started at 2 and extended to 10, and another started at 6 and extended to 12. We could translate (or slide) each of these to place their midpoint at zero by subtracting the mean value from each of their endpoints. For the first specimen, we would subtract 6, translating the specimen from (2,10) to (-4,4), and for the second we would subtract 9, tanslating the specimen from (6,12) to (-3,3).</p>
<p><em>Rotation</em> is an angular rotation to bring a 2 or 3D object into a common orientation. For a more thorough discussion please see <span class="citation" data-cites="Zelditch:2004">(<a href="#ref-Zelditch:2004" role="doc-biblioref">Zelditch 2004</a>)</span>.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="../../images/Procrustes_superimposition.png" class="img-fluid figure-img"></p>
<p></p><figcaption class="figure-caption">Figure: Procrustes superimposition. The figure shows the three transformation steps of an ordinary Procrustes fit for two configurations of landmarks. (a) Scaling of both configurations to the same size; (b) Translation to the same position of the center of gravity; (c) Rotation to the orientation that provides the minimum sum of squared distances between corresponding landmarks.</figcaption><p></p>
</figure>
</div>
<p><a href="https://commons.wikimedia.org/wiki/File:Procrustes_superimposition.png">Source: Christian Klingenberg</a></p>
<div class="callout-note callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
From the documentation for <code>gpagen()</code> in <code>geomorph</code>:
</div>
</div>
<div class="callout-body-container callout-body">
<p>GPA translates all specimens to the origin, scales them to unit-centroid size, and optimally rotates them (using a least-squares criterion) until the coordinates of corresponding points align as closely as possible. The resulting aligned Procrustes coordinates represent the shape of each specimen, and are found in a curved space related to Kendall’s shape space (Kendall 1984). Typically, these are projected into a linear tangent space yielding Kendall’s tangent space coordinates (i.e., Procrustes shape variables), which are used for subsequent multivariate analyses (Dryden and Mardia 1993, Rohlf 1999).</p>
</div>
</div>
</section>
</section>
<section id="geomorph-package-1" class="level1">
<h1>Geomorph Package</h1>
<section id="data-structures" class="level3">
<h3 class="anchored" data-anchor-id="data-structures">Data Structures</h3>
<p>The package <code>geomorph</code> is written to work with standard geomorphic morphometrics data file types such as TPS and NTS, as well as others.</p>
<p><code>readland.tps()</code> reads morphometric data in TPS format. It returns an array for storing landmark data: <em>p</em> landmarks X <em>k</em> dimensions X <em>N</em> specimens</p>
<p>The third dimension of this array contains names for each specimen, which are obtained from the image names in the <em>.tps</em> file.</p>
</section>
<section id="built-in-example" class="level3">
<h3 class="anchored" data-anchor-id="built-in-example">Built-in Example</h3>
<p>Letʻs look again at the built-in dataset <code>plethodon</code> that comes with <code>geomorph</code>. It has skull landmark data for 40 specimens of two species of plethdon salamanders. <code>plethodon</code> is a list.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(plethodon) </span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(plethodon)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "list"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="fu">attributes</span>(plethodon)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$names
[1] "land"    "links"   "species" "site"    "outline"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(plethodon<span class="sc">$</span>species)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
 Jord Teyah 
   20    20 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(plethodon<span class="sc">$</span>land)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 12  2 40</code></pre>
</div>
</div>
<p>The landmarks are in <code>plethodon$land</code>. There are twenty of each species, and from the dimensions of <code>plethodon$land</code> we see that each of the 40 specimens has 12 landmarks in two dimensions. <code>plotAllSpecimens</code> plots the coordinates of all samples, so we can start to look at variation.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="co">#|warning: false</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plotAllSpecimens</span>(plethodon<span class="sc">$</span>land)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="index_files/figure-html/unnamed-chunk-6-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>Y.gpa <span class="ot">&lt;-</span> <span class="fu">gpagen</span>(plethodon<span class="sc">$</span>land, <span class="at">PrinAxes =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Performing GPA

  |                                                                            
  |                                                                      |   0%
  |                                                                            
  |==================                                                    |  25%
  |                                                                            
  |===================================                                   |  50%
  |                                                                            
  |======================================================================| 100%

Making projections... Finished!</code></pre>
</div>
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(Y.gpa)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
gpagen(A = plethodon$land, PrinAxes = FALSE) 



Generalized Procrustes Analysis
with Partial Procrustes Superimposition

12 fixed landmarks
0 semilandmarks (sliders)
2-dimensional landmarks
2 GPA iterations to converge


Consensus (mean) Configuration

             X            Y
1   0.15263287 -0.023350400
2   0.19445064 -0.092643121
3  -0.03361223 -0.007346212
4  -0.28069721 -0.092850147
5  -0.30998751 -0.061672264
6  -0.32557841 -0.036112255
7  -0.31804390  0.036125318
8  -0.18947114  0.098010902
9   0.02036829  0.099112877
10  0.18852641  0.077278061
11  0.35134314  0.065866346
12  0.55006905 -0.062419105</code></pre>
</div>
</div>
<p>The summary provides the mean for each of the coordinates after GPA. We can see the mean of the coordinates among the individual coordinates by plotting:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(Y.gpa)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="index_files/figure-html/unnamed-chunk-7-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Letʻs compare the plot of all specimens before and after Generalized Procrustes Analysis. The below plot is before GPA, on the raw landmarks <code>plethodon$land</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotAllSpecimens</span>(plethodon<span class="sc">$</span>land)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="index_files/figure-html/unnamed-chunk-8-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Plotting again after alignment with Generalized Procrustes Analysis gives a good representation of shape variation. Plotting links between the landmarks helps with visualization.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotAllSpecimens</span>(Y.gpa<span class="sc">$</span>coords,<span class="at">links=</span>plethodon<span class="sc">$</span>links)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="index_files/figure-html/unnamed-chunk-9-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>PCA on the procrustes coordinates (size-adjusted or shape variables) can be done with:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>pleth.pca <span class="ot">&lt;-</span> <span class="fu">gm.prcomp</span>(Y.gpa<span class="sc">$</span>coords)</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>pleth.pca </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Ordination type: Principal Component Analysis 
Centering by OLS mean
Orthogonal projection of OLS residuals
Number of observations: 40 
Number of vectors 20 

Importance of Components:
                             Comp1       Comp2        Comp3        Comp4
Eigenvalues            0.001855441 0.001566597 0.0004141056 0.0002278444
Proportion of Variance 0.367433295 0.310233333 0.0820053761 0.0451200584
Cumulative Proportion  0.367433295 0.677666628 0.7596720044 0.8047920628
                              Comp5        Comp6        Comp7        Comp8
Eigenvalues            0.0001725997 0.0001672575 0.0001549008 0.0001251132
Proportion of Variance 0.0341799312 0.0331220169 0.0306750315 0.0247761773
Cumulative Proportion  0.8389719940 0.8720940109 0.9027690424 0.9275452197
                              Comp9       Comp10       Comp11       Comp12
Eigenvalues            8.468503e-05 6.869926e-05 5.430853e-05 4.423148e-05
Proportion of Variance 1.677019e-02 1.360452e-02 1.075473e-02 8.759165e-03
Cumulative Proportion  9.443154e-01 9.579199e-01 9.686747e-01 9.774338e-01
                             Comp13       Comp14       Comp15       Comp16
Eigenvalues            0.0000261742 0.0000192644 1.741769e-05 1.715141e-05
Proportion of Variance 0.0051832795 0.0038149309 3.449226e-03 3.396495e-03
Cumulative Proportion  0.9826170994 0.9864320303 9.898813e-01 9.932778e-01
                             Comp17       Comp18       Comp19       Comp20
Eigenvalues            1.555244e-05 9.253679e-06 5.336746e-06 3.802714e-06
Proportion of Variance 3.079853e-03 1.832507e-03 1.056837e-03 7.530520e-04
Cumulative Proportion  9.963576e-01 9.981901e-01 9.992469e-01 1.000000e+00</code></pre>
</div>
</div>
<p>There are 20 PC axes, with 67% of the total shape variation in the first two axes. A plot on PC1 vs PC2 reveals several distinct clusters, suggesting shape differences. The scores are in <code>pleth.pca$x</code>. There is a plot method for geomorph pca results:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>pleth.pca.plot <span class="ot">&lt;-</span> <span class="fu">plot</span>(pleth.pca)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="index_files/figure-html/unnamed-chunk-11-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>We can plot additional by specifying the axes to plot:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">2</span>,<span class="dv">2</span>)) <span class="co"># set up 2x2 plots</span></span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(pleth.pca, <span class="at">main =</span> <span class="st">"PCA"</span>)</span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(pleth.pca, <span class="at">main =</span> <span class="st">"PCA"</span>, <span class="at">axis1 =</span> <span class="dv">1</span>, <span class="at">axis2 =</span> <span class="dv">3</span>) </span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a>          <span class="co"># change PCs viewed to axes 1 vs 3</span></span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(pleth.pca, <span class="at">main =</span> <span class="st">"PCA"</span>, <span class="at">axis1 =</span> <span class="dv">2</span>, <span class="at">axis2 =</span> <span class="dv">3</span>) </span>
<span id="cb30-6"><a href="#cb30-6" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(pleth.pca, <span class="at">main =</span> <span class="st">"PCA"</span>, <span class="at">axis1 =</span> <span class="dv">3</span>, <span class="at">axis2 =</span> <span class="dv">4</span>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="index_files/figure-html/unnamed-chunk-12-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>The specimens are from two species in two distinct environments, and we can test these as explanatory variables for phenotypic varation. Because there are multiple pheontypic (Y) variables, we use MANOVA (multivariate ANOVA). reveals sig- nificant shape differences between species, between sites and in the interaction between the two factors:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>y<span class="ot">&lt;-</span><span class="fu">two.d.array</span>(Y.gpa<span class="sc">$</span>coords) </span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>lm.fit <span class="ot">&lt;-</span> <span class="fu">procD.lm</span>(y<span class="sc">~</span>plethodon<span class="sc">$</span>species<span class="sc">*</span>plethodon<span class="sc">$</span>site, <span class="at">iter=</span><span class="dv">99</span>)</span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(lm.fit)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Analysis of Variance, using Residual Randomization
Permutation procedure: Randomization of null model residuals 
Number of permutations: 100 
Estimation method: Ordinary Least Squares 
Sums of Squares and Cross-products: Type I 
Effect sizes (Z) based on F distributions

                                 Df       SS       MS     Rsq      F      Z
plethodon$species                 1 0.029258 0.029258 0.14856 14.544 3.9999
plethodon$site                    1 0.064375 0.064375 0.32688 32.000 6.2349
plethodon$species:plethodon$site  1 0.030885 0.030885 0.15682 15.352 5.3284
Residuals                        36 0.072422 0.002012 0.36774              
Total                            39 0.196940                               
                                 Pr(&gt;F)  
plethodon$species                  0.01 *
plethodon$site                     0.01 *
plethodon$species:plethodon$site   0.01 *
Residuals                                
Total                                    
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Call: procD.lm(f1 = y ~ plethodon$species * plethodon$site, iter = 99)</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>ref<span class="ot">&lt;-</span><span class="fu">mshape</span>(Y.gpa<span class="sc">$</span>coords)</span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>gp1.mn<span class="ot">&lt;-</span><span class="fu">mshape</span>(Y.gpa<span class="sc">$</span>coords[,,<span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>])</span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plotRefToTarget</span>(ref,gp1.mn,<span class="at">mag=</span><span class="dv">2</span>, <span class="at">links=</span>plethodon<span class="sc">$</span>links)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="index_files/figure-html/unnamed-chunk-14-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
</section>
<section id="more-software" class="level1">
<h1>More software</h1>
<p><a href="https://academic.oup.com/view-large/223239151" class="uri">https://academic.oup.com/view-large/223239151</a></p>



</section>

<div id="quarto-appendix" class="default"><section class="quarto-appendix-contents" role="doc-bibliography"><h2 class="anchored quarto-appendix-heading">References</h2><div id="refs" class="references csl-bib-body hanging-indent" role="doc-bibliography">
<div id="ref-Zelditch:2004" class="csl-entry" role="doc-biblioentry">
Zelditch, Miriam. 2004. <em>Geometric Morphometrics for Biologists a Primer</em>. Amsterdam ; Elsevier Academic Press.
</div>
</div></section></div></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>