<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.2.313">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Marguerite Butler">
<meta name="dcterms.date" content="2023-04-25">
<meta name="description" content="Working with trees with tidytree and plotting with ggtree">

<title>ZOOL710 Data Science in R for Biologists - The ggtree-verse</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>


<link rel="stylesheet" href="../../styles.css">
</head>

<body class="floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">ZOOL710 Data Science in R for Biologists</span>
    </a>
  </div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../index.html">
 <span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-general-information" role="button" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">General Information</span>
    </a>
    <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="nav-menu-general-information">    
        <li>
    <a class="dropdown-item" href="../../syllabus.html">
 <span class="dropdown-text">Syllabus</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../schedule.html">
 <span class="dropdown-text">Schedule</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-course-materials" role="button" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Course Materials</span>
    </a>
    <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="nav-menu-course-materials">    
        <li>
    <a class="dropdown-item" href="../../lectures.html">
 <span class="dropdown-text">Lectures</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../projects.html">
 <span class="dropdown-text">Projects</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../resources.html">
 <span class="dropdown-text">Resources</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/"><i class="bi bi-github" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://twitter.com"><i class="bi bi-twitter" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="../../index.xml"><i class="bi bi-rss" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
</ul>
              <div id="quarto-search" class="" title="Search"></div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<header id="title-block-header" class="quarto-title-block default toc-left page-columns page-full">
  <div class="quarto-title-banner page-columns page-full">
    <div class="quarto-title column-body">
      <h1 class="title">The ggtree-verse</h1>
                  <div>
        <div class="description">
          Working with trees with tidytree and plotting with ggtree
        </div>
      </div>
                          <div class="quarto-categories">
                <div class="quarto-category">module 7</div>
                <div class="quarto-category">week 13</div>
                <div class="quarto-category">phylogenetic trees</div>
                <div class="quarto-category">phylo</div>
                <div class="quarto-category">ggtree</div>
                <div class="quarto-category">treedata</div>
              </div>
                  </div>
  </div>
    
  <div class="quarto-title-meta-author">
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-heading">Affiliation</div>
    
      <div class="quarto-title-meta-contents">
      <a href="https://butlerlab.org">Marguerite Butler</a> 
    </div>
      <div class="quarto-title-meta-contents">
          <p class="affiliation">
              School of Life Sciences, University of Hawaii
            </p>
        </div>
      </div>

  <div class="quarto-title-meta">

        
      <div>
      <div class="quarto-title-meta-heading">Published</div>
      <div class="quarto-title-meta-contents">
        <p class="date">April 25, 2023</p>
      </div>
    </div>
    
      
    </div>
    
  
  </header><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse sidebar-navigation floating overflow-auto">
    <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#acknowledgements" id="toc-acknowledgements" class="nav-link active" data-scroll-target="#acknowledgements">Acknowledgements</a></li>
  <li><a href="#learning-objectives" id="toc-learning-objectives" class="nav-link" data-scroll-target="#learning-objectives">Learning objectives</a></li>
  <li><a href="#overview" id="toc-overview" class="nav-link" data-scroll-target="#overview">Overview</a></li>
  <li><a href="#phylo---treedata" id="toc-phylo---treedata" class="nav-link" data-scroll-target="#phylo---treedata">phylo -&gt; treedata</a>
  <ul class="collapse">
  <li><a href="#coverting-between-data-types" id="toc-coverting-between-data-types" class="nav-link" data-scroll-target="#coverting-between-data-types">Coverting between data types</a></li>
  </ul></li>
  <li><a href="#simulate-phenotypic-data" id="toc-simulate-phenotypic-data" class="nav-link" data-scroll-target="#simulate-phenotypic-data">Simulate phenotypic data</a>
  <ul class="collapse">
  <li><a href="#add-tiplabels" id="toc-add-tiplabels" class="nav-link" data-scroll-target="#add-tiplabels">Add tiplabels</a></li>
  <li><a href="#get-the-tip-labels" id="toc-get-the-tip-labels" class="nav-link" data-scroll-target="#get-the-tip-labels">Get the tip labels</a></li>
  </ul></li>
  <li><a href="#merge-tree-with-data" id="toc-merge-tree-with-data" class="nav-link" data-scroll-target="#merge-tree-with-data">Merge tree with data</a>
  <ul class="collapse">
  <li><a href="#subsetting-the-tree" id="toc-subsetting-the-tree" class="nav-link" data-scroll-target="#subsetting-the-tree">Subsetting the tree</a></li>
  <li><a href="#plotting-with-node-labels" id="toc-plotting-with-node-labels" class="nav-link" data-scroll-target="#plotting-with-node-labels">Plotting with node labels</a></li>
  <li><a href="#plotting-with-alternative-tip-labels" id="toc-plotting-with-alternative-tip-labels" class="nav-link" data-scroll-target="#plotting-with-alternative-tip-labels">Plotting with alternative tip labels</a></li>
  <li><a href="#when-your-tip-labels-get-cut-off" id="toc-when-your-tip-labels-get-cut-off" class="nav-link" data-scroll-target="#when-your-tip-labels-get-cut-off">When your tip labels get cut off</a></li>
  </ul></li>
  <li><a href="#tree-layouts" id="toc-tree-layouts" class="nav-link" data-scroll-target="#tree-layouts">Tree layouts</a></li>
  <li><a href="#plotting-data-on-the-tree" id="toc-plotting-data-on-the-tree" class="nav-link" data-scroll-target="#plotting-data-on-the-tree">Plotting data on the tree</a>
  <ul class="collapse">
  <li><a href="#example-plot-smalltree-with-size-in-a-barplot" id="toc-example-plot-smalltree-with-size-in-a-barplot" class="nav-link" data-scroll-target="#example-plot-smalltree-with-size-in-a-barplot">Example: plot <code>smalltree</code> with <code>size</code> in a barplot</a></li>
  </ul></li>
  <li><a href="#a-tree-panel-and-annotation-example-from-the-tree-data-book" id="toc-a-tree-panel-and-annotation-example-from-the-tree-data-book" class="nav-link" data-scroll-target="#a-tree-panel-and-annotation-example-from-the-tree-data-book">A tree-panel and annotation example from the Tree Data Book:</a>
  <ul class="collapse">
  <li><a href="#the-operator-for-ggtree-objects" id="toc-the-operator-for-ggtree-objects" class="nav-link" data-scroll-target="#the-operator-for-ggtree-objects">The %+&gt;% operator for ggtree objects</a></li>
  <li><a href="#example-of-the-operator-to-add-data-to-a-ggtree-object." id="toc-example-of-the-operator-to-add-data-to-a-ggtree-object." class="nav-link" data-scroll-target="#example-of-the-operator-to-add-data-to-a-ggtree-object.">Example of the %+&gt;% operator to add data to a ggtree object.</a>
  <ul class="collapse">
  <li><a href="#explore-df_info" id="toc-explore-df_info" class="nav-link" data-scroll-target="#explore-df_info">Explore <code>df_info</code></a></li>
  <li><a href="#add-snp-and-trait-plots-aligned-to-the-tree" id="toc-add-snp-and-trait-plots-aligned-to-the-tree" class="nav-link" data-scroll-target="#add-snp-and-trait-plots-aligned-to-the-tree">Add SNP and Trait plots aligned to the tree</a></li>
  </ul></li>
  </ul></li>
  <li><a href="#example-datasets" id="toc-example-datasets" class="nav-link" data-scroll-target="#example-datasets">Example datasets</a></li>
  </ul>
</nav>
</nav>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
    </div>
<!-- main -->
<main class="content quarto-banner-title-block" id="quarto-document-content">




<section id="acknowledgements" class="level3">
<h3 class="anchored" data-anchor-id="acknowledgements">Acknowledgements</h3>
<p>References for this Material:</p>
<ul>
<li>Treedata book by Guangchuang Yu <span class="citation" data-cites="Yu:2022">(<a href="#ref-Yu:2022" role="doc-biblioref">Yu 2022</a>)</span> <a href="https://yulab-smu.top/treedata-book/" class="uri">https://yulab-smu.top/treedata-book/</a></li>
<li><a href="https://bioconnector.github.io/workshops/r-ggtree.html" class="uri">https://bioconnector.github.io/workshops/r-ggtree.html</a></li>
</ul>
</section>
<section id="learning-objectives" class="level1">
<h1>Learning objectives</h1>
<div class="callout-note callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Learning Objectives
</div>
</div>
<div class="callout-body-container callout-body">
<p><strong>At the end of this lesson you will:</strong></p>
<ul>
<li>Understand information content of phylogenetically structured data</li>
<li>Understand particular R tree formats in ape, phylobase, and ouch</li>
<li>Be able to hand-make trees</li>
<li>Be able to import trees from nexus, newick, and other major formats in use today</li>
<li>Be able to convert trees from one format to another</li>
<li>Be able to perform basic tree manipulations</li>
</ul>
</div>
</div>
<div class="callout-note callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
R packages you will need
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li><code>ape</code></li>
<li><code>ggtree</code></li>
<li><code>tidytree</code></li>
<li><code>treeio</code></li>
</ul>
</div>
</div>
</section>
<section id="overview" class="level1">
<h1>Overview</h1>
<p><code>ggtree</code> is a powerful phylogenetic tree plotting package, that is used alongside <code>treeio</code> and <code>tidytree</code> packages to import/export and to manipulate phylogenetic trees, respectively. These packages allow you to subset or combine trees with data, annotate, and plot in so many different ways using the grammar of graphics.</p>
<p>Because itʻs written in the style of <code>ggplot2</code> and the <code>tidyverse</code>, there are a lot of functions to remember, and now there are also multiple object types. The key to working effectively with <code>ggtree</code> is to be aware of what type of object you are working with, whether it is a <code>dataframe</code>, a <code>tibble</code>, a <code>phylo</code> tree, a <code>treedata</code> tibble, or <code>ggtree</code> object. It is easy to convert between these at will if you know what you are dealing with.</p>
</section>
<section id="phylo---treedata" class="level1">
<h1>phylo -&gt; treedata</h1>
<p>Functions like <code>read.tree</code> amd <code>read.nexus</code>, etc. will read in objects of class <code>phylo</code> (they are actually referencing the ape function).</p>
<p>To show this, letʻs first generate a random tree using <code>ape::rtree()</code></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">require</span>(ape)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">require</span>(ggtree)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">require</span>(treeio)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>tree <span class="ot">&lt;-</span> <span class="fu">rtree</span>(<span class="dv">20</span>)</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(tree)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "phylo"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(tree)  <span class="co"># ape plotting function</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="index_files/figure-html/unnamed-chunk-2-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Note: <code>ggtree</code> can also accept <code>phylo</code> objects as arguments:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggtree</span>(tree)  <span class="co"># ggtree plotting function</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="index_files/figure-html/unnamed-chunk-3-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>To save as newick and nexus formats, <code>treeio</code> has the following:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>treeio<span class="sc">::</span><span class="fu">write.nexus</span>(tree, <span class="at">file=</span><span class="st">"tree.nex"</span>)</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>treeio<span class="sc">::</span><span class="fu">write.tree</span>(tree, <span class="at">file=</span><span class="st">"tree.tree"</span>)</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a><span class="fu">list.files</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "_index.qmd"         "anolis.SSD.raw.csv" "bigtree.nex"       
 [4] "ggtree_functions.R" "ggtree.R"           "index_files"       
 [7] "index.qmd"          "index.rmarkdown"    "tree.nex"          
[10] "tree.tree"         </code></pre>
</div>
</div>
<p>Take a look at these files. They should be in nexus and newick formats, respectively.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>tree_nex <span class="ot">&lt;-</span> treeio<span class="sc">::</span><span class="fu">read.nexus</span>(<span class="at">file=</span><span class="st">"tree.nex"</span>)</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>tree_new <span class="ot">&lt;-</span> treeio<span class="sc">::</span><span class="fu">read.newick</span>(<span class="at">file=</span><span class="st">"tree.tree"</span>)</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(tree_nex)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "phylo"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(tree_new)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "phylo"</code></pre>
</div>
</div>
<div class="callout-note callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>If you have trees from <code>iqtree</code>, <code>BEAST</code> etc., you should use the specialized read functions because they will capture the metadata embedded in the treeformat including support values.</p>
</div>
</div>
<section id="coverting-between-data-types" class="level2">
<h2 class="anchored" data-anchor-id="coverting-between-data-types">Coverting between data types</h2>
<p>The main functions for coersion are:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">as.phylo</span> (tree)  <span class="co"># to phylo</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Phylogenetic tree with 20 tips and 19 internal nodes.

Tip labels:
  t20, t10, t3, t7, t6, t18, ...

Rooted; includes branch lengths.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="fu">as.treedata</span>(tree) <span class="co"># to treedata</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>'treedata' S4 object'.

...@ phylo:

Phylogenetic tree with 20 tips and 19 internal nodes.

Tip labels:
  t20, t10, t3, t7, t6, t18, ...

Rooted; includes branch lengths.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>treedata <span class="ot">&lt;-</span> <span class="fu">as.treedata</span>(tree)  </span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a><span class="fu">as_tibble</span>(treedata)  <span class="co"># to tibble</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 39 × 4
   parent  node branch.length label
    &lt;int&gt; &lt;int&gt;         &lt;dbl&gt; &lt;chr&gt;
 1     22     1       0.915   t20  
 2     26     2       0.532   t10  
 3     26     3       0.847   t3   
 4     25     4       0.118   t7   
 5     28     5       0.690   t6   
 6     28     6       0.725   t18  
 7     29     7       0.261   t14  
 8     31     8       0.00314 t9   
 9     31     9       0.0128  t4   
10     30    10       0.289   t12  
# … with 29 more rows
# ℹ Use `print(n = ...)` to see more rows</code></pre>
</div>
</div>
</section>
</section>
<section id="simulate-phenotypic-data" class="level1">
<h1>Simulate phenotypic data</h1>
<section id="add-tiplabels" class="level3">
<h3 class="anchored" data-anchor-id="add-tiplabels">Add tiplabels</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggtree</span>(tree) <span class="sc">+</span> <span class="fu">geom_tiplab</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="index_files/figure-html/unnamed-chunk-7-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="get-the-tip-labels" class="level3">
<h3 class="anchored" data-anchor-id="get-the-tip-labels">Get the tip labels</h3>
<p>There is a function <code>get_taxa_name()</code> which is very handy, but NOTE: its argument is a <code>ggtree</code> object, not a <code>treedata</code> object.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">ggtree</span>(tree) <span class="sc">+</span> <span class="fu">geom_tiplab</span>()</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a><span class="fu">get_taxa_name</span>(p)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "t4"  "t9"  "t12" "t14" "t18" "t6"  "t3"  "t10" "t7"  "t17" "t15" "t13"
[13] "t11" "t5"  "t2"  "t16" "t20" "t1"  "t8"  "t19"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>taxa <span class="ot">&lt;-</span> <span class="fu">get_taxa_name</span>(p)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Simulate a data matrix:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>n <span class="ot">&lt;-</span> <span class="fu">length</span>(taxa)</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>size <span class="ot">&lt;-</span> <span class="fu">rnorm</span>(n, <span class="at">mean=</span><span class="dv">20</span>, <span class="at">sd=</span><span class="dv">5</span>)</span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>habitat <span class="ot">&lt;-</span> <span class="fu">sample</span>(<span class="fu">c</span>(<span class="st">"desert"</span>, <span class="st">"grassland"</span>, <span class="st">"forest"</span>, <span class="st">"intertidal"</span>), <span class="at">size=</span>n, <span class="at">replace=</span>T)</span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a>dat <span class="ot">&lt;-</span> <span class="fu">data.frame</span>( <span class="st">"label"</span><span class="ot">=</span> taxa, size, habitat)</span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a>dat</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   label      size    habitat
1     t4 13.190731     desert
2     t9 19.541849  grassland
3    t12 24.957978 intertidal
4    t14 14.831695     forest
5    t18 21.221452     desert
6     t6 10.671196 intertidal
7     t3 15.308488 intertidal
8    t10 19.347808     desert
9     t7 21.088887     forest
10   t17  7.394603     desert
11   t15 14.215988 intertidal
12   t13 11.341547     desert
13   t11 17.769369 intertidal
14    t5 13.667986  grassland
15    t2 18.265933  grassland
16   t16 18.950921     desert
17   t20 21.236842 intertidal
18    t1 17.947431  grassland
19    t8 11.629983 intertidal
20   t19 13.964597     forest</code></pre>
</div>
</div>
<p>See our ggtree as a treedata object:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="fu">as.treedata</span>(p) <span class="sc">%&gt;%</span> as_tibble <span class="sc">%&gt;%</span> as.data.frame  </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   parent node branch.length label
1      22    1   0.914660785   t20
2      26    2   0.531915339   t10
3      26    3   0.847355583    t3
4      25    4   0.117541116    t7
5      28    5   0.689902873    t6
6      28    6   0.725284685   t18
7      29    7   0.260759907   t14
8      31    8   0.003144709    t9
9      31    9   0.012793490    t4
10     30   10   0.289318340   t12
11     33   11   0.102822345   t16
12     33   12   0.361748627    t2
13     34   13   0.406217780    t5
14     36   14   0.855577735   t13
15     37   15   0.122123984   t15
16     37   16   0.055801166   t17
17     35   17   0.411075242   t11
18     38   18   0.613953796   t19
19     39   19   0.453259011    t8
20     39   20   0.470733995    t1
21     21   21            NA  &lt;NA&gt;
22     21   22   0.768506521  &lt;NA&gt;
23     22   23   0.153267556  &lt;NA&gt;
24     23   24   0.854931734  &lt;NA&gt;
25     24   25   0.451706708  &lt;NA&gt;
26     25   26   0.354669122  &lt;NA&gt;
27     24   27   0.472259514  &lt;NA&gt;
28     27   28   0.512345967  &lt;NA&gt;
29     27   29   0.681458147  &lt;NA&gt;
30     29   30   0.226860373  &lt;NA&gt;
31     30   31   0.391176049  &lt;NA&gt;
32     23   32   0.034940638  &lt;NA&gt;
33     32   33   0.130967953  &lt;NA&gt;
34     32   34   0.677282662  &lt;NA&gt;
35     34   35   0.356319352  &lt;NA&gt;
36     35   36   0.147734449  &lt;NA&gt;
37     36   37   0.980167842  &lt;NA&gt;
38     21   38   0.107984108  &lt;NA&gt;
39     38   39   0.106352382  &lt;NA&gt;</code></pre>
</div>
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>  <span class="co"># ggtree -&gt; treedata -&gt; tibble -&gt; dataframe</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="merge-tree-with-data" class="level1">
<h1>Merge tree with data</h1>
<p>Now that we have a matching key in both the tree and data objects, we can join the tree with the dataframe by those matching labels using <code>ggtree</code>ʻs <code>full_join</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>ttree <span class="ot">&lt;-</span> <span class="fu">full_join</span>(tree, dat, <span class="at">by =</span> <span class="st">"label"</span>) </span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>ttree</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>'treedata' S4 object'.

...@ phylo:

Phylogenetic tree with 20 tips and 19 internal nodes.

Tip labels:
  t20, t10, t3, t7, t6, t18, ...

Rooted; includes branch lengths.

with the following features available:
  'size', 'habitat'.

# The associated data tibble abstraction: 39 × 5
# The 'node', 'label' and 'isTip' are from the phylo tree.
    node label isTip  size habitat   
   &lt;int&gt; &lt;chr&gt; &lt;lgl&gt; &lt;dbl&gt; &lt;chr&gt;     
 1     1 t20   TRUE   21.2 intertidal
 2     2 t10   TRUE   19.3 desert    
 3     3 t3    TRUE   15.3 intertidal
 4     4 t7    TRUE   21.1 forest    
 5     5 t6    TRUE   10.7 intertidal
 6     6 t18   TRUE   21.2 desert    
 7     7 t14   TRUE   14.8 forest    
 8     8 t9    TRUE   19.5 grassland 
 9     9 t4    TRUE   13.2 desert    
10    10 t12   TRUE   25.0 intertidal
# … with 29 more rows
# ℹ Use `print(n = ...)` to see more rows</code></pre>
</div>
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>ttree <span class="sc">%&gt;%</span> as_tibble <span class="sc">%&gt;%</span> as.data.frame</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   parent node branch.length label      size    habitat
1      22    1   0.914660785   t20 21.236842 intertidal
2      26    2   0.531915339   t10 19.347808     desert
3      26    3   0.847355583    t3 15.308488 intertidal
4      25    4   0.117541116    t7 21.088887     forest
5      28    5   0.689902873    t6 10.671196 intertidal
6      28    6   0.725284685   t18 21.221452     desert
7      29    7   0.260759907   t14 14.831695     forest
8      31    8   0.003144709    t9 19.541849  grassland
9      31    9   0.012793490    t4 13.190731     desert
10     30   10   0.289318340   t12 24.957978 intertidal
11     33   11   0.102822345   t16 18.950921     desert
12     33   12   0.361748627    t2 18.265933  grassland
13     34   13   0.406217780    t5 13.667986  grassland
14     36   14   0.855577735   t13 11.341547     desert
15     37   15   0.122123984   t15 14.215988 intertidal
16     37   16   0.055801166   t17  7.394603     desert
17     35   17   0.411075242   t11 17.769369 intertidal
18     38   18   0.613953796   t19 13.964597     forest
19     39   19   0.453259011    t8 11.629983 intertidal
20     39   20   0.470733995    t1 17.947431  grassland
21     21   21            NA  &lt;NA&gt;        NA       &lt;NA&gt;
22     21   22   0.768506521  &lt;NA&gt;        NA       &lt;NA&gt;
23     22   23   0.153267556  &lt;NA&gt;        NA       &lt;NA&gt;
24     23   24   0.854931734  &lt;NA&gt;        NA       &lt;NA&gt;
25     24   25   0.451706708  &lt;NA&gt;        NA       &lt;NA&gt;
26     25   26   0.354669122  &lt;NA&gt;        NA       &lt;NA&gt;
27     24   27   0.472259514  &lt;NA&gt;        NA       &lt;NA&gt;
28     27   28   0.512345967  &lt;NA&gt;        NA       &lt;NA&gt;
29     27   29   0.681458147  &lt;NA&gt;        NA       &lt;NA&gt;
30     29   30   0.226860373  &lt;NA&gt;        NA       &lt;NA&gt;
31     30   31   0.391176049  &lt;NA&gt;        NA       &lt;NA&gt;
32     23   32   0.034940638  &lt;NA&gt;        NA       &lt;NA&gt;
33     32   33   0.130967953  &lt;NA&gt;        NA       &lt;NA&gt;
34     32   34   0.677282662  &lt;NA&gt;        NA       &lt;NA&gt;
35     34   35   0.356319352  &lt;NA&gt;        NA       &lt;NA&gt;
36     35   36   0.147734449  &lt;NA&gt;        NA       &lt;NA&gt;
37     36   37   0.980167842  &lt;NA&gt;        NA       &lt;NA&gt;
38     21   38   0.107984108  &lt;NA&gt;        NA       &lt;NA&gt;
39     38   39   0.106352382  &lt;NA&gt;        NA       &lt;NA&gt;</code></pre>
</div>
</div>
<p>And thatʻs what our <code>treedata</code> object looks like flattened out!</p>
<section id="subsetting-the-tree" class="level2">
<h2 class="anchored" data-anchor-id="subsetting-the-tree">Subsetting the tree</h2>
<p>Functions: <code>drop.tip()</code> and <code>keep.tip()</code></p>
<p>Suppose we want to drop all of the even tips:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>todrop <span class="ot">&lt;-</span> <span class="fu">paste</span>(<span class="st">"t"</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span><span class="sc">*</span><span class="dv">2</span>, <span class="at">sep=</span><span class="st">""</span>)</span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>todrop</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "t2"  "t4"  "t6"  "t8"  "t10" "t12" "t14" "t16" "t18" "t20"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>smalltree <span class="ot">&lt;-</span> <span class="fu">drop.tip</span>(ttree, todrop)</span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>smalltree</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>'treedata' S4 object'.

...@ phylo:

Phylogenetic tree with 10 tips and 9 internal nodes.

Tip labels:
  t3, t7, t9, t5, t13, t15, ...

Rooted; includes branch lengths.

with the following features available:
  'size', 'habitat'.

# The associated data tibble abstraction: 19 × 5
# The 'node', 'label' and 'isTip' are from the phylo tree.
    node label isTip  size habitat   
   &lt;int&gt; &lt;chr&gt; &lt;lgl&gt; &lt;dbl&gt; &lt;chr&gt;     
 1     1 t3    TRUE  15.3  intertidal
 2     2 t7    TRUE  21.1  forest    
 3     3 t9    TRUE  19.5  grassland 
 4     4 t5    TRUE  13.7  grassland 
 5     5 t13   TRUE  11.3  desert    
 6     6 t15   TRUE  14.2  intertidal
 7     7 t17   TRUE   7.39 desert    
 8     8 t11   TRUE  17.8  intertidal
 9     9 t19   TRUE  14.0  forest    
10    10 t1    TRUE  17.9  grassland 
# … with 9 more rows
# ℹ Use `print(n = ...)` to see more rows</code></pre>
</div>
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggtree</span>(smalltree) <span class="sc">+</span> <span class="fu">geom_tiplab</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="index_files/figure-html/unnamed-chunk-12-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p><code>drop.tip</code> keeps all of the metadata! <code>keep.tip</code> is imported from <code>ape</code> so it has to be converted to <code>phylo</code> and then the data joined again after.</p>
</section>
<section id="plotting-with-node-labels" class="level2">
<h2 class="anchored" data-anchor-id="plotting-with-node-labels">Plotting with node labels</h2>
<p>The geometries <code>geom_text()</code> and <code>geom_node()</code> are helpful for labelling all of the nodes. The function <code>geom_tiplab()</code> labels only the tips.</p>
<p>Add node labels so you know what the internal node numbers are:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggtree</span>(smalltree) <span class="sc">+</span> </span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_tiplab</span>() <span class="sc">+</span></span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_text</span>(<span class="fu">aes</span>(<span class="at">label=</span>node), <span class="at">hjust=</span><span class="sc">-</span>.<span class="dv">3</span>)    <span class="co"># node numbers</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="index_files/figure-html/unnamed-chunk-13-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Note: The tiplabels and the node labels crashed!</p>
<p>There are also <code>2</code> versions: <code>geom_text2()</code> and <code>geom_node2()</code> that allow subsetting the nodes, when you want the geometry to apply to only some of the nodes.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggtree</span>(smalltree) <span class="sc">+</span> </span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_tiplab</span>() <span class="sc">+</span></span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_text2</span>(<span class="fu">aes</span>(<span class="at">label=</span>node, <span class="at">subset=</span><span class="sc">!</span>isTip), <span class="at">hjust=</span><span class="sc">-</span>.<span class="dv">3</span>)    <span class="co"># node numbers</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="index_files/figure-html/unnamed-chunk-14-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p><code>isTip</code> is a column of the ggtree object, so it is inherited when we provide the ggtree object.</p>
</section>
<section id="plotting-with-alternative-tip-labels" class="level2">
<h2 class="anchored" data-anchor-id="plotting-with-alternative-tip-labels">Plotting with alternative tip labels</h2>
<p>The dataframe portion of the treedata object can hold any number of columns of metadata. Perhaps you have some real names in a different column (like a display name), it is easy to swap out the tip labels. Here letʻs just use the <code>habitat</code> column</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggtree</span>(smalltree) <span class="sc">+</span> </span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_tiplab</span>(<span class="fu">aes</span>(<span class="at">label=</span>habitat)) <span class="sc">+</span></span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_text2</span>(<span class="fu">aes</span>(<span class="at">label=</span>node, <span class="at">subset=</span><span class="sc">!</span>isTip), <span class="at">hjust=</span><span class="sc">-</span>.<span class="dv">3</span>)    <span class="co"># node numbers</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="index_files/figure-html/unnamed-chunk-15-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="when-your-tip-labels-get-cut-off" class="level2">
<h2 class="anchored" data-anchor-id="when-your-tip-labels-get-cut-off">When your tip labels get cut off</h2>
<p>Add an x scale (usually time):</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggtree</span>(smalltree) <span class="sc">+</span> </span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_tiplab</span>(<span class="fu">aes</span>(<span class="at">label=</span>habitat)) <span class="sc">+</span></span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_text2</span>(<span class="fu">aes</span>(<span class="at">label=</span>node, <span class="at">subset=</span><span class="sc">!</span>isTip), <span class="at">hjust=</span><span class="sc">-</span>.<span class="dv">3</span>) <span class="sc">+</span>   <span class="co"># node numbers</span></span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_tree2</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="index_files/figure-html/unnamed-chunk-16-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>You can increase the size of the plot area to accommodate the longer labels:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggtree</span>(smalltree) <span class="sc">+</span> </span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_tiplab</span>(<span class="fu">aes</span>(<span class="at">label=</span>habitat)) <span class="sc">+</span></span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_text2</span>(<span class="fu">aes</span>(<span class="at">label=</span>node, <span class="at">subset=</span><span class="sc">!</span>isTip), <span class="at">hjust=</span><span class="sc">-</span>.<span class="dv">3</span>) <span class="sc">+</span>   <span class="co"># node numbers</span></span>
<span id="cb40-4"><a href="#cb40-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_tree2</span>() <span class="sc">+</span></span>
<span id="cb40-5"><a href="#cb40-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">xlim</span>(<span class="dv">0</span>,<span class="dv">5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="index_files/figure-html/unnamed-chunk-17-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
</section>
<section id="tree-layouts" class="level1">
<h1>Tree layouts</h1>
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="fu">require</span>(cowplot)</span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_grid</span>(</span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtree</span>(ttree),</span>
<span id="cb41-5"><a href="#cb41-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtree</span>(ttree, <span class="at">branch.length=</span><span class="st">'none'</span>),</span>
<span id="cb41-6"><a href="#cb41-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtree</span>(ttree, <span class="at">layout=</span><span class="st">"dendrogram"</span>),</span>
<span id="cb41-7"><a href="#cb41-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtree</span>(ttree, <span class="at">layout=</span><span class="st">"roundrect"</span>),</span>
<span id="cb41-8"><a href="#cb41-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtree</span>(ttree, <span class="at">layout=</span><span class="st">"ellipse"</span>),</span>
<span id="cb41-9"><a href="#cb41-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtree</span>(ttree, <span class="at">layout=</span><span class="st">"ellipse"</span>, <span class="at">branch.length=</span><span class="st">"none"</span>),</span>
<span id="cb41-10"><a href="#cb41-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtree</span>(ttree, <span class="at">layout=</span><span class="st">"circular"</span>),</span>
<span id="cb41-11"><a href="#cb41-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtree</span>(ttree, <span class="at">branch.length=</span><span class="st">'none'</span>, <span class="at">layout=</span><span class="st">'circular'</span>),</span>
<span id="cb41-12"><a href="#cb41-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtree</span>(ttree, <span class="at">layout=</span><span class="st">"fan"</span>, <span class="at">open.angle=</span><span class="dv">120</span>),</span>
<span id="cb41-13"><a href="#cb41-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtree</span>(ttree, <span class="at">layout=</span><span class="st">"inward_circular"</span>)</span>
<span id="cb41-14"><a href="#cb41-14" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="plotting-data-on-the-tree" class="level1">
<h1>Plotting data on the tree</h1>
<p><code>geom_facet()</code> and <code>facet_plot()</code> are general methods to link graphical layers to a tree.</p>
<p>These functions require an input dataframe with the first column containing the taxon labels (the key which matches to the tip labels of the phylogeny).</p>
<p>Internally these functions reorder the input data based on the tree structure so that you donʻt have to worry about the order of the rows.</p>
<p>Multiple layers can be added to the same dataset. Also different datasets can be added to the same figure.</p>
<p>A table of the <strong>geom layers</strong> that work with <code>geom_facet</code> is <a href="https://yulab-smu.top/treedata-book/figures-and-tables.html#tab:facet-geom">provided here</a>.</p>
<section id="example-plot-smalltree-with-size-in-a-barplot" class="level2">
<h2 class="anchored" data-anchor-id="example-plot-smalltree-with-size-in-a-barplot">Example: plot <code>smalltree</code> with <code>size</code> in a barplot</h2>
<p>First make a tibble to attach to the tree. As of this writing, <code>geom_facet</code> will not accept a <code>treedata</code> object. It wants a dataframe or tibble of only the tips. But this is easy to make from the treedata. We just have to filter out the non-tip rows, then rearrange the columns to put the labels first:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="fu">require</span>(ggplot2)</span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a><span class="fu">require</span>(dplyr)</span>
<span id="cb42-3"><a href="#cb42-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-4"><a href="#cb42-4" aria-hidden="true" tabindex="-1"></a>smdat <span class="ot">&lt;-</span> smalltree <span class="sc">%&gt;%</span> </span>
<span id="cb42-5"><a href="#cb42-5" aria-hidden="true" tabindex="-1"></a>           as_tibble <span class="sc">%&gt;%</span> </span>
<span id="cb42-6"><a href="#cb42-6" aria-hidden="true" tabindex="-1"></a>           <span class="fu">filter</span>(<span class="sc">!</span><span class="fu">is.na</span>(label)) <span class="sc">%&gt;%</span></span>
<span id="cb42-7"><a href="#cb42-7" aria-hidden="true" tabindex="-1"></a>           <span class="fu">relocate</span>(label, <span class="at">.before=</span><span class="dv">1</span>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can then add the barplot as a panel next to the tree plot using <code>geom_facet</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggtree</span>(smalltree) <span class="sc">+</span> </span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_tiplab</span>() <span class="sc">+</span></span>
<span id="cb43-3"><a href="#cb43-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_tree2</span>()  <span class="sc">+</span> </span>
<span id="cb43-4"><a href="#cb43-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_facet</span>(<span class="at">panel =</span> <span class="st">"Size"</span>, </span>
<span id="cb43-5"><a href="#cb43-5" aria-hidden="true" tabindex="-1"></a>               <span class="at">data=</span>smdat, </span>
<span id="cb43-6"><a href="#cb43-6" aria-hidden="true" tabindex="-1"></a>               <span class="at">geom =</span> geom_col, </span>
<span id="cb43-7"><a href="#cb43-7" aria-hidden="true" tabindex="-1"></a>               <span class="at">mapping=</span><span class="fu">aes</span>(<span class="at">x =</span> smdat<span class="sc">$</span>size), </span>
<span id="cb43-8"><a href="#cb43-8" aria-hidden="true" tabindex="-1"></a>               <span class="at">orientation =</span> <span class="st">'y'</span>, </span>
<span id="cb43-9"><a href="#cb43-9" aria-hidden="true" tabindex="-1"></a>               <span class="at">width =</span> .<span class="dv">6</span>, </span>
<span id="cb43-10"><a href="#cb43-10" aria-hidden="true" tabindex="-1"></a>               <span class="at">fill=</span><span class="st">"blue"</span>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="index_files/figure-html/unnamed-chunk-20-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>The arguments for geom_facet() are:</p>
<ul>
<li><code>panel</code> : The name of the panel, displayed on top</li>
<li><code>data</code> : a <code>tibble</code> or <code>dataframe</code> containing the metadata to plot. Must have as the first column the tip labels that are found in the phylogenetic tree.</li>
<li><code>geom</code> : a <a href="https://yulab-smu.top/treedata-book/figures-and-tables.html#tab:facet-geom">geom layer</a> specifying the style of plot</li>
<li><code>mapping</code> : the aesthetic mapping. I should not have to supply the <code>smdat$</code> here but it wonʻt work otherwise.</li>
<li>any additional parameters for the plot</li>
</ul>
</section>
</section>
<section id="a-tree-panel-and-annotation-example-from-the-tree-data-book" class="level1">
<h1>A tree-panel and annotation example from the Tree Data Book:</h1>
<p>This example plots a phylogeny alongside SNP (single nucleotide polymorphism) data and a barplot of some simulated data <span class="citation" data-cites="Yu:2022">(<a href="#ref-Yu:2022" role="doc-biblioref">Yu 2022</a>)</span>.</p>
<section id="the-operator-for-ggtree-objects" class="level2">
<h2 class="anchored" data-anchor-id="the-operator-for-ggtree-objects">The %+&gt;% operator for ggtree objects</h2>
<p>The %+&gt;% operator is used to add data (<code>dataframe</code>, <code>tibble</code>) to a <code>ggtree</code> object:</p>
<div class="sourceCode" id="cb44"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="ex">my_ggtree</span> <span class="op">&lt;</span>- my_ggtree %<span class="op">&lt;</span>+% new_data</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>The result is a combined object that can be used for plotting, but it does not modify the original <code>treedata</code> object (which is a different object from the <code>ggtree</code> object). The <code>full_join()</code> function can be used to combine a tree with data to produce a new <code>treedata</code> object.</p>
</section>
<section id="example-of-the-operator-to-add-data-to-a-ggtree-object." class="level2">
<h2 class="anchored" data-anchor-id="example-of-the-operator-to-add-data-to-a-ggtree-object.">Example of the %+&gt;% operator to add data to a ggtree object.</h2>
<p>The package <code>TDbook</code> is the data accompanyment to <span class="citation" data-cites="Yu:2022">(<a href="#ref-Yu:2022" role="doc-biblioref">Yu 2022</a>)</span>ʻs <a href="https://yulab-smu.top/treedata-book/index.html">Tree Data book</a>. It is available on CRAN so you can install it with the usual <code>install.packages("TDbook")</code> function call.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="fu">require</span>(ggtree)</span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a><span class="fu">require</span>(ggplot2)</span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a><span class="fu">require</span>(dplyr)</span>
<span id="cb45-4"><a href="#cb45-4" aria-hidden="true" tabindex="-1"></a><span class="fu">require</span>(TDbook)</span>
<span id="cb45-5"><a href="#cb45-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-6"><a href="#cb45-6" aria-hidden="true" tabindex="-1"></a><span class="co"># load `tree_boots`, `df_tip_data`, and `df_inode_data` from 'TDbook'</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">ggtree</span>(tree_boots) <span class="sc">%&lt;+%</span> df_tip_data <span class="sc">+</span> <span class="fu">xlim</span>(<span class="sc">-</span>.<span class="dv">1</span>, <span class="dv">4</span>)</span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">geom_tiplab</span>(<span class="at">offset =</span> .<span class="dv">6</span>, <span class="at">hjust =</span> .<span class="dv">5</span>) <span class="sc">+</span></span>
<span id="cb46-3"><a href="#cb46-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_tippoint</span>(<span class="fu">aes</span>(<span class="at">shape =</span> trophic_habit, <span class="at">color =</span> trophic_habit, </span>
<span id="cb46-4"><a href="#cb46-4" aria-hidden="true" tabindex="-1"></a>                <span class="at">size =</span> mass_in_kg)) <span class="sc">+</span> </span>
<span id="cb46-5"><a href="#cb46-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"right"</span>) <span class="sc">+</span> </span>
<span id="cb46-6"><a href="#cb46-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_size_continuous</span>(<span class="at">range =</span> <span class="fu">c</span>(<span class="dv">3</span>, <span class="dv">10</span>))</span>
<span id="cb46-7"><a href="#cb46-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb46-8"><a href="#cb46-8" aria-hidden="true" tabindex="-1"></a>p2 <span class="sc">%&lt;+%</span> df_inode_data <span class="sc">+</span> </span>
<span id="cb46-9"><a href="#cb46-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_label</span>(<span class="fu">aes</span>(<span class="at">label =</span> vernacularName.y, <span class="at">fill =</span> posterior)) <span class="sc">+</span> </span>
<span id="cb46-10"><a href="#cb46-10" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_fill_gradientn</span>(<span class="at">colors =</span> RColorBrewer<span class="sc">::</span><span class="fu">brewer.pal</span>(<span class="dv">3</span>, <span class="st">"YlGnBu"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Removed 7 rows containing missing values (`geom_label()`).</code></pre>
</div>
<div class="cell-output-display">
<p><img src="index_files/figure-html/unnamed-chunk-22-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<section id="explore-df_info" class="level3">
<h3 class="anchored" data-anchor-id="explore-df_info">Explore <code>df_info</code></h3>
<p><code>df_info</code> A dataframe containing sampling info for the tips of the tree. 386 rows and 6 variables, with the first column being taxa labels (<code>id</code>).</p>
<p><code>df_alleles</code> The allele table with original raw data to be processed to SNP data. It is a table of nucleotides with 386 rows x 385 variables. The first row contains tips labels. Column names are non-sense. The rownames (exept for the first one) contains the snp position along the genome.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a><span class="do">## load `tree_nwk`, `df_info`, `df_alleles`, and `df_bar_data` from 'TDbook'</span></span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a>tree <span class="ot">&lt;-</span> tree_nwk</span>
<span id="cb48-3"><a href="#cb48-3" aria-hidden="true" tabindex="-1"></a>snps <span class="ot">&lt;-</span> df_alleles</span>
<span id="cb48-4"><a href="#cb48-4" aria-hidden="true" tabindex="-1"></a>snps_strainCols <span class="ot">&lt;-</span> snps[<span class="dv">1</span>,] </span>
<span id="cb48-5"><a href="#cb48-5" aria-hidden="true" tabindex="-1"></a>snps<span class="ot">&lt;-</span>snps[<span class="sc">-</span><span class="dv">1</span>,] <span class="co"># drop strain names</span></span>
<span id="cb48-6"><a href="#cb48-6" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(snps) <span class="ot">&lt;-</span> snps_strainCols</span>
<span id="cb48-7"><a href="#cb48-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb48-8"><a href="#cb48-8" aria-hidden="true" tabindex="-1"></a>gapChar <span class="ot">&lt;-</span> <span class="st">"?"</span></span>
<span id="cb48-9"><a href="#cb48-9" aria-hidden="true" tabindex="-1"></a>snp <span class="ot">&lt;-</span> <span class="fu">t</span>(snps)  <span class="co"># not rectangular!</span></span>
<span id="cb48-10"><a href="#cb48-10" aria-hidden="true" tabindex="-1"></a>lsnp <span class="ot">&lt;-</span> <span class="fu">apply</span>(snp, <span class="dv">1</span>, <span class="cf">function</span>(x) {</span>
<span id="cb48-11"><a href="#cb48-11" aria-hidden="true" tabindex="-1"></a>        x <span class="sc">!=</span> snp[<span class="dv">1</span>,] <span class="sc">&amp;</span> x <span class="sc">!=</span> gapChar <span class="sc">&amp;</span> snp[<span class="dv">1</span>,] <span class="sc">!=</span> gapChar</span>
<span id="cb48-12"><a href="#cb48-12" aria-hidden="true" tabindex="-1"></a>    })  <span class="co"># different from row 1, not missing</span></span>
<span id="cb48-13"><a href="#cb48-13" aria-hidden="true" tabindex="-1"></a>lsnp <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(lsnp) </span>
<span id="cb48-14"><a href="#cb48-14" aria-hidden="true" tabindex="-1"></a>lsnp<span class="sc">$</span>pos <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(<span class="fu">rownames</span>(lsnp))  <span class="co"># position from rownames</span></span>
<span id="cb48-15"><a href="#cb48-15" aria-hidden="true" tabindex="-1"></a>lsnp <span class="ot">&lt;-</span> tidyr<span class="sc">::</span><span class="fu">gather</span>(lsnp, name, value, <span class="sc">-</span>pos)  </span>
<span id="cb48-16"><a href="#cb48-16" aria-hidden="true" tabindex="-1"></a>snp_data <span class="ot">&lt;-</span> lsnp[lsnp<span class="sc">$</span>value, <span class="fu">c</span>(<span class="st">"name"</span>, <span class="st">"pos"</span>)] <span class="co"># only TRUEs</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><code>snp_data</code> A dataframe containing SNP position data. 6482 x 2. The first column contains taxa labels coresponding to the tips of the tree (<code>name</code>). There are multiple rows per taxon, the second colum is the position <code>pos</code> of the snp in the genome. This is used as the x-variable in the plot.</p>
<p>In the dataframe <code>snp_data</code> the rows are ordered by position along the sequence (the x-dimension of this data), but the first column is the strain (taxon) name which matches the tips in the phylogenetic tree.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a><span class="do">## visualize the tree </span></span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">ggtree</span>(tree) </span>
<span id="cb49-3"><a href="#cb49-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-4"><a href="#cb49-4" aria-hidden="true" tabindex="-1"></a><span class="do">## attach the sampling information data set </span></span>
<span id="cb49-5"><a href="#cb49-5" aria-hidden="true" tabindex="-1"></a><span class="do">## and add symbols colored by location</span></span>
<span id="cb49-6"><a href="#cb49-6" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">%&lt;+%</span> df_info <span class="sc">+</span> <span class="fu">geom_tippoint</span>(<span class="fu">aes</span>(<span class="at">color=</span>location))</span>
<span id="cb49-7"><a href="#cb49-7" aria-hidden="true" tabindex="-1"></a>p</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="index_files/figure-html/unnamed-chunk-24-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="add-snp-and-trait-plots-aligned-to-the-tree" class="level3">
<h3 class="anchored" data-anchor-id="add-snp-and-trait-plots-aligned-to-the-tree">Add SNP and Trait plots aligned to the tree</h3>
<p>Use <code>geom_facet</code> with reference to the respective dataframes/tibbles to add plots alignted to the tree. For the SNP plot, we will use <code>geom_point</code> which allows x-y plotting, with x-coordinate determined by <code>pos</code> and the y-coordinate aligned with the taxon. The symbol is the vertical line <code>|</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a><span class="do">## visualize SNP and Trait data using dot and bar charts,</span></span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a><span class="do">## and align them based on tree structure</span></span>
<span id="cb50-3"><a href="#cb50-3" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">geom_facet</span>(<span class="at">panel =</span> <span class="st">"SNP"</span>, <span class="at">data =</span> snp_data, <span class="at">geom =</span> geom_point, </span>
<span id="cb50-4"><a href="#cb50-4" aria-hidden="true" tabindex="-1"></a>               <span class="at">mapping=</span><span class="fu">aes</span>(<span class="at">x =</span> pos, <span class="at">color =</span> location), <span class="at">shape =</span> <span class="st">'|'</span>) </span>
<span id="cb50-5"><a href="#cb50-5" aria-hidden="true" tabindex="-1"></a>p1</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="index_files/figure-html/unnamed-chunk-25-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p><code>df_bar_data</code> is some simulated data with an <code>id</code> column specifying the taxon names, and a <code>dummy_bar_value</code> containing some data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a>p1 <span class="sc">+</span> <span class="fu">geom_facet</span>(<span class="at">panel =</span> <span class="st">"Trait"</span>, <span class="at">data =</span> df_bar_data, <span class="at">geom =</span> geom_col, </span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a>                <span class="fu">aes</span>(<span class="at">x =</span> dummy_bar_value, <span class="at">color =</span> location, </span>
<span id="cb51-3"><a href="#cb51-3" aria-hidden="true" tabindex="-1"></a>                <span class="at">fill =</span> location), <span class="at">orientation =</span> <span class="st">'y'</span>, <span class="at">width =</span> .<span class="dv">6</span>) <span class="sc">+</span></span>
<span id="cb51-4"><a href="#cb51-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_tree2</span>(<span class="at">legend.position=</span><span class="fu">c</span>(.<span class="dv">05</span>, .<span class="dv">85</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="index_files/figure-html/unnamed-chunk-26-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
</section>
</section>
<section id="example-datasets" class="level1">
<h1>Example datasets</h1>
<p>save to your working directory:</p>
<p><a href="bigtree.nex">bigtree.nex</a><br>
<a href="anolis.SSD.raw.csv">anolis.SSD.raw.csv</a><br>
<a href="ggtree.R">ggtree.R</a></p>
<p>This is an example of a typical workflow. We have carefully collected phenotypic data, and someone has published a massive phylogeny. We need to subset the tree to just the taxa we want to work on.</p>



</section>

<div id="quarto-appendix" class="default"><section class="quarto-appendix-contents" role="doc-bibliography"><h2 class="anchored quarto-appendix-heading">References</h2><div id="refs" class="references csl-bib-body hanging-indent" role="doc-bibliography">
<div id="ref-Yu:2022" class="csl-entry" role="doc-biblioentry">
Yu, G. 2022. <em>Data Integration, Manipulation and Visualization of Phylogenetic Trees</em>. A Chapman &amp; Hall Book. CRC Press. <a href="https://yulab-smu.top/treedata-book/">https://yulab-smu.top/treedata-book/</a>.
</div>
</div></section></div></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>