<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.2.313">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Marguerite Butler">
<meta name="dcterms.date" content="2023-04-18">
<meta name="description" content="Getting trees in to R and understanding their formats">

<title>ZOOL710 Data Science in R for Biologists - All about trees</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>

  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<link rel="stylesheet" href="../../styles.css">
</head>

<body class="floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">ZOOL710 Data Science in R for Biologists</span>
    </a>
  </div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../index.html">
 <span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-general-information" role="button" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">General Information</span>
    </a>
    <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="nav-menu-general-information">    
        <li>
    <a class="dropdown-item" href="../../syllabus.html">
 <span class="dropdown-text">Syllabus</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../schedule.html">
 <span class="dropdown-text">Schedule</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-course-materials" role="button" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Course Materials</span>
    </a>
    <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="nav-menu-course-materials">    
        <li>
    <a class="dropdown-item" href="../../lectures.html">
 <span class="dropdown-text">Lectures</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../projects.html">
 <span class="dropdown-text">Projects</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../resources.html">
 <span class="dropdown-text">Resources</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/"><i class="bi bi-github" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://twitter.com"><i class="bi bi-twitter" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="../../index.xml"><i class="bi bi-rss" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
</ul>
              <div id="quarto-search" class="" title="Search"></div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<header id="title-block-header" class="quarto-title-block default toc-left page-columns page-full">
  <div class="quarto-title-banner page-columns page-full">
    <div class="quarto-title column-body">
      <h1 class="title">All about trees</h1>
                  <div>
        <div class="description">
          Getting trees in to R and understanding their formats
        </div>
      </div>
                          <div class="quarto-categories">
                <div class="quarto-category">module 7</div>
                <div class="quarto-category">week 13</div>
                <div class="quarto-category">phylogenetic trees</div>
                <div class="quarto-category">nexus</div>
                <div class="quarto-category">fasta</div>
                <div class="quarto-category">newick</div>
                <div class="quarto-category">beast</div>
                <div class="quarto-category">iqtree</div>
              </div>
                  </div>
  </div>
    
  <div class="quarto-title-meta-author">
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-heading">Affiliation</div>
    
      <div class="quarto-title-meta-contents">
      <a href="https://butlerlab.org">Marguerite Butler</a> 
    </div>
      <div class="quarto-title-meta-contents">
          <p class="affiliation">
              School of Life Sciences, University of Hawaii
            </p>
        </div>
      </div>

  <div class="quarto-title-meta">

        
      <div>
      <div class="quarto-title-meta-heading">Published</div>
      <div class="quarto-title-meta-contents">
        <p class="date">April 18, 2023</p>
      </div>
    </div>
    
      
    </div>
    
  
  </header><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse sidebar-navigation floating overflow-auto">
    <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#acknowledgements" id="toc-acknowledgements" class="nav-link active" data-scroll-target="#acknowledgements">Acknowledgements</a></li>
  <li><a href="#learning-objectives" id="toc-learning-objectives" class="nav-link" data-scroll-target="#learning-objectives">Learning objectives</a></li>
  <li><a href="#tree-vocabulary" id="toc-tree-vocabulary" class="nav-link" data-scroll-target="#tree-vocabulary">Tree vocabulary</a></li>
  <li><a href="#newick-format" id="toc-newick-format" class="nav-link" data-scroll-target="#newick-format">Newick format</a></li>
  <li><a href="#nexus-more-soon" id="toc-nexus-more-soon" class="nav-link" data-scroll-target="#nexus-more-soon">NEXUS – More soon</a>
  <ul class="collapse">
  <li><a href="#nexus-syntax" id="toc-nexus-syntax" class="nav-link" data-scroll-target="#nexus-syntax">Nexus syntax:</a></li>
  <li><a href="#note-partition-files-in-iqtree" id="toc-note-partition-files-in-iqtree" class="nav-link" data-scroll-target="#note-partition-files-in-iqtree">Note: Partition files in IQTREE</a></li>
  </ul></li>
  <li><a href="#phylo-ape-5.0-or-above" id="toc-phylo-ape-5.0-or-above" class="nav-link" data-scroll-target="#phylo-ape-5.0-or-above"><code>phylo</code> (<code>ape</code> 5.0 or above)</a>
  <ul class="collapse">
  <li><a href="#getting-trees-into-phylo-format" id="toc-getting-trees-into-phylo-format" class="nav-link" data-scroll-target="#getting-trees-into-phylo-format">Getting trees into phylo format</a></li>
  </ul></li>
  <li><a href="#ouchtree-ouch-version-2.0-or-above" id="toc-ouchtree-ouch-version-2.0-or-above" class="nav-link" data-scroll-target="#ouchtree-ouch-version-2.0-or-above">ouchtree (<code>ouch</code> version 2.0 or above)</a>
  <ul class="collapse">
  <li><a href="#building-ouchtree-format-from-dataframes" id="toc-building-ouchtree-format-from-dataframes" class="nav-link" data-scroll-target="#building-ouchtree-format-from-dataframes">Building ouchtree format from dataframes</a></li>
  <li><a href="#converting-from-ape-to-ouch" id="toc-converting-from-ape-to-ouch" class="nav-link" data-scroll-target="#converting-from-ape-to-ouch">Converting from <code>ape</code> to <code>ouch</code></a></li>
  <li><a href="#coerce-to-data.frame" id="toc-coerce-to-data.frame" class="nav-link" data-scroll-target="#coerce-to-data.frame">Coerce to data.frame</a></li>
  <li><a href="#hidden-components-of-ouchtree" id="toc-hidden-components-of-ouchtree" class="nav-link" data-scroll-target="#hidden-components-of-ouchtree">Hidden components of <code>ouchtree</code></a></li>
  </ul></li>
  <li><a href="#treedata-from-package-tidytree" id="toc-treedata-from-package-tidytree" class="nav-link" data-scroll-target="#treedata-from-package-tidytree"><code>treedata</code> from package <code>tidytree</code></a>
  <ul class="collapse">
  <li><a href="#reading-trees-into-treedata" id="toc-reading-trees-into-treedata" class="nav-link" data-scroll-target="#reading-trees-into-treedata">Reading trees into <code>treedata</code></a></li>
  <li><a href="#combining-tree-with-data" id="toc-combining-tree-with-data" class="nav-link" data-scroll-target="#combining-tree-with-data">Combining tree with data</a></li>
  <li><a href="#methods-to-extract-information-from-treedata" id="toc-methods-to-extract-information-from-treedata" class="nav-link" data-scroll-target="#methods-to-extract-information-from-treedata">Methods to extract information from <code>treedata</code></a></li>
  <li><a href="#plotting-ggtree" id="toc-plotting-ggtree" class="nav-link" data-scroll-target="#plotting-ggtree">Plotting ggtree</a></li>
  </ul></li>
  </ul>
</nav>
</nav>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
    </div>
<!-- main -->
<main class="content quarto-banner-title-block" id="quarto-document-content">




<section id="acknowledgements" class="level3">
<h3 class="anchored" data-anchor-id="acknowledgements">Acknowledgements</h3>
<p>References for this Material:</p>
<ul>
<li>This was adapted from the chapter “All about trees” written by Brian OʻMeara from the manual “Getting started in R for Phylogenetics” by Marguerite A. Butler, Brian C. O’Meara, Jason Pienaar, Michael Alfaro, Graham Slater, and Todd Oakley</li>
</ul>
</section>
<section id="learning-objectives" class="level1">
<h1>Learning objectives</h1>
<div class="callout-note callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Learning Objectives
</div>
</div>
<div class="callout-body-container callout-body">
<p><strong>At the end of this lesson you will:</strong></p>
<ul>
<li>Understand information content of phylogenetically structured data</li>
<li>Understand particular R tree formats in ape, phylobase, and ouch</li>
<li>Be able to hand-make trees</li>
<li>Be able to import trees from nexus, newick, and other major formats in use today</li>
<li>Be able to convert trees from one format to another</li>
<li>Be able to perform basic tree manipulations</li>
</ul>
</div>
</div>
<div class="callout-note callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
R packages you will need
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li><code>ape</code></li>
<li><code>ouch</code></li>
<li><code>ggtree</code> which will also install <code>treeio</code>, <code>tidytree</code>, and others</li>
</ul>
<p><code>ggtree</code> is distributed through <code>bioconductor</code>. * To install ggtree:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">require</span>(<span class="st">"BiocManager"</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>))</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">install.packages</span>(<span class="st">"BiocManager"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Bioconductor version '3.15' is out-of-date; the current release version '3.16'
  is available with R version '4.2'; see https://bioconductor.org/install</code></pre>
</div>
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">"ggtree"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Bioconductor version 3.15 (BiocManager 1.30.18), R 4.2.1 (2022-06-23)</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Installing package(s) 'ggtree'</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>
The downloaded binary packages are in
    /var/folders/zg/q7cjrc5j4f91jcc089mwhvvm0000gn/T//Rtmplqg0gm/downloaded_packages</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Old packages: 'ade4', 'aplot', 'BiocManager', 'bit', 'blob', 'boot', 'brew',
  'broom', 'bslib', 'cachem', 'callr', 'car', 'class', 'cli', 'cluster',
  'codetools', 'colorspace', 'commonmark', 'covr', 'cpp11', 'crayon', 'curl',
  'data.table', 'dbplyr', 'desc', 'deSolve', 'devtools', 'digest', 'dotCall64',
  'dplyr', 'dtplyr', 'evaluate', 'expm', 'fansi', 'fastmap', 'fontawesome',
  'forcats', 'foreign', 'fs', 'gapminder', 'gargle', 'gee', 'geiger', 'GenSA',
  'geodist', 'geometry', 'gert', 'ggbeeswarm', 'ggfun', 'ggimage', 'ggplot2',
  'ggrepel', 'gh', 'gitcreds', 'googledrive', 'googlesheets4', 'gtable',
  'gtools', 'haven', 'highr', 'hms', 'httpuv', 'httr', 'igraph', 'isoband',
  'jsonlite', 'knitr', 'lattice', 'lifecycle', 'lme4', 'lpSolve', 'lubridate',
  'magic', 'magick', 'maps', 'markdown', 'MASS', 'Matrix', 'mgcv', 'minqa',
  'mnormt', 'modelr', 'nlme', 'nnet', 'openssl', 'parameters', 'phangorn',
  'phytools', 'pillar', 'pkgbuild', 'pkgdown', 'pkgload', 'plyr', 'processx',
  'ps', 'purrr', 'quantreg', 'ragg', 'RcppArmadillo', 'readr', 'readxl',
  'rlang', 'RMySQL', 'rncl', 'RNeXML', 'roxygen2', 'rpart', 'RPostgreSQL',
  'RSQLite', 'rstudioapi', 'rversions', 'sass', 'scales', 'scatterplot3d',
  'segmented', 'seqinr', 'shiny', 'sourcetools', 'sp', 'spatial', 'spelling',
  'statmod', 'stringi', 'stringr', 'survival', 'svglite', 'sys', 'terra',
  'testthat', 'tibble', 'tidyr', 'tidyselect', 'tidytree', 'tidyverse',
  'tinytex', 'treeio', 'utf8', 'V8', 'vctrs', 'vdiffr', 'vegan', 'viridisLite',
  'vroom', 'webfakes', 'whisker', 'xfun', 'XML', 'yaml', 'yulab.utils', 'zip'</code></pre>
</div>
</div>
</div>
</div>
</section>
<section id="tree-vocabulary" class="level1">
<h1>Tree vocabulary</h1>
<p>In nature, a tree is a large perennial woody plant. It has roots, a main trunk, branches, and leaves. In graph theory, a tree is a network where there is only one path between any two nodes (in other words, a network with no loops or cycles).</p>
<p>In phylogenetics, we use ideas and terminology from both graph theory and nature:</p>
<ul>
<li><strong>Terminal taxa</strong> are also known as <strong>leaves</strong>, <strong>terminals</strong>, <strong>OTU</strong>s (“<strong>Operational Taxonomic Units</strong>”), <strong>tips</strong>, or simply <strong>taxa</strong>.</li>
<li><strong>Branches</strong> are also called <strong>edges</strong> or <strong>internodes</strong>.</li>
<li><strong>Internal</strong> nodes (places where two or more branches connect) are also known as <strong>vertices</strong> and sometimes simply <strong>nodes</strong> (technically, leaves are also nodes).</li>
<li>A <strong>rooted</strong> tree has one node designated as the <strong>root</strong>, and all other nodes are descended from this root. Traditionally, the root node has at least two descendants; it may also have a subtending branch (an ancestral branch).</li>
<li>An <strong>unrooted</strong> tree has no root designated, making the oldest ancestor ambiguous.<br>
</li>
<li>A tree where every internal node has two and only two descendants is known as a <strong>binary</strong> or <strong>bifurcating</strong> tree.</li>
<li>A tree where at least one internal node has more than two descendants is said to be <strong>multifurcating</strong>; such a node is a <strong>polytomy</strong>.</li>
<li>Trees in phylogenetics generally represent either <strong>species trees</strong> (a history of the splitting of interbreeding populations) or <strong>gene trees</strong> (a history of the coalescence of gene copies). In both cases, it is generally believed that the true process is bifurcating, so that each split results in two descendants. Thus, polytomies on trees are generally taken as representing <em>uncertainty in the relationships</em>. However in the case of adaptive radiations, and other <strong>rapidly speciating processes</strong>, the true process may be approaching multifurcation.</li>
<li><strong>Branches</strong> may have <strong>lengths</strong>; these lengths may correspond to <em>time</em>, <em>amount of change</em> (e.g., mutations) in some set of characters, <em>number of speciation events</em>, or some other measure.</li>
<li>A tree where all <strong>branch lengths</strong> from root to tips are equal is known as an <strong>ultrametric</strong> tree.</li>
<li>A tree without branch lengths is known as a <strong>topology</strong>.</li>
<li>A <strong>monophyletic clade</strong> is an ancestor and all its descendants.</li>
<li>Any <strong>edge</strong> or <strong>branch</strong> corresponds to a <strong>bipartition</strong>: <em>a division of the tree into two parts connected by that edge</em> (if a root were inserted on the edge, then each of those parts would be a clade).</li>
</ul>
</section>
<section id="newick-format" class="level1">
<h1>Newick format</h1>
<p>A very basic tree format is <strong>Newick</strong> (<code>.tree</code>), named after the <a href="https://www.newicks.com">seafood restaurant</a> in New Hampshire where it was formalized in 1986 by James Archie, William H. E. Day, Joseph Felsenstein, Wayne Maddison, Christopher Meacham, F. James Rohlf, and David Swofford - the second meeting was at the restaurant; it is also called New Hampshire format for that reason). It is widely used in graph theory as well. It is simply a string. Each nesting on the tree corresponds to a parenthetical statement. For example, for this tree (drawn brute-force):</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="http://www.newicks.com/Web2014/Ed7.1.jpg" class="img-fluid figure-img"></p>
<p></p><figcaption class="figure-caption">Photo from the Newick restaurant <a href="https://www.newicks.com">home page</a></figcaption><p></p>
</figure>
</div>
<p>NOTE: You would never construct a <code>phylo</code> tree as below, this is just to illustrate its structure (see following sections for how to read in trees):</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">require</span>(ape)</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="fu">require</span>(ouch)</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="fu">require</span>(treeio)</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a><span class="co"># simpletree &lt;- rcoal(length(LETTERS[1:7]), rooted = TRUE, tip.label = LETTERS[1:7], br = "coalescent")</span></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>simpletree<span class="ot">&lt;-</span><span class="fu">structure</span>(<span class="fu">list</span>(</span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">edge =</span> <span class="fu">matrix</span>(</span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a>      <span class="fu">c</span>(<span class="dv">8</span>, <span class="dv">9</span>, <span class="dv">10</span>, <span class="dv">10</span>, <span class="dv">9</span>, <span class="dv">11</span>, <span class="dv">11</span>, <span class="dv">8</span>, <span class="dv">12</span>, <span class="dv">12</span>, <span class="dv">13</span>, <span class="dv">13</span>, </span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a>        <span class="dv">9</span>, <span class="dv">10</span>, <span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">11</span>, <span class="dv">3</span>, <span class="dv">4</span>, <span class="dv">12</span>, <span class="dv">5</span>, <span class="dv">13</span>, <span class="dv">6</span>, <span class="dv">7</span>),</span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a>      <span class="at">nrow=</span>12L), </span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a>    <span class="at">tip.label =</span> <span class="fu">c</span>(<span class="st">"A"</span>, <span class="st">"B"</span>, <span class="st">"C"</span>, <span class="st">"D"</span>, <span class="st">"E"</span>, <span class="st">"F"</span>, <span class="st">"G"</span>), </span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true" tabindex="-1"></a>    <span class="at">Nnode =</span> 6L</span>
<span id="cb8-13"><a href="#cb8-13" aria-hidden="true" tabindex="-1"></a>  ), </span>
<span id="cb8-14"><a href="#cb8-14" aria-hidden="true" tabindex="-1"></a>    <span class="at">class =</span> <span class="st">"phylo"</span>)</span>
<span id="cb8-15"><a href="#cb8-15" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(simpletree,<span class="at">no.margin=</span><span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div id="fig-simpletree" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="index_files/figure-html/fig-simpletree-1.png" class="img-fluid figure-img" width="672"></p>
<p></p><figcaption class="figure-caption">Figure&nbsp;1: A simple tree</figcaption><p></p>
</figure>
</div>
</div>
</div>
<p>Taxa G and F form a <strong>clade</strong>, as do G, F, and E, as do A and B, and so forth. Thus, to create a Newick string, just go down the tree, nesting as you go:</p>
<p><strong>(G,F)</strong></p>
<p><strong>((G,F),E)</strong></p>
<p>other side:</p>
<p><strong>(A,B)</strong></p>
<p><strong>(C,D)</strong></p>
<p><strong>((A,B),(C,D))</strong></p>
<p>all together:</p>
<p><strong>(((G,F),E),((A,B),(C,D)))</strong></p>
<p>And that’s it (it will be clearer in the lecture) If a tree has branch lengths, these are entered following the descendant clade. For example, if the branch leading to G has length 1.0, we would write <code>G:1.0</code> rather than just <code>G</code>. If the tree is ultrametric, and the branch below the common ancestor of G and F is of length <code>1.1</code>, and the branch below that of length <code>3.5</code>, we could write (note: a semicolon <code>;</code> ends the newick string):</p>
<p><code>((G:1.0,F:1.0):1.1,E:2.1):3.5;</code></p>
<p>One aspect of the Newick (and most tree) representation is that there are many ways of representing the same tree. At every node, one can rotate the descendant branches (switching the left and right positions) and get the same tree (for example, imagine switching the G and E labels). Thus, the Newick strings</p>
<p><strong>((G,F),E)</strong></p>
<p>and</p>
<p><strong>((F,G),E)</strong></p>
<p>describe the same trees, though it might not be easy to tell at first glance. This is generally an issue for any tree representation. While Newick strings are compact and easy to understand, they also don’t lend themselves to easy <em>tree traversal</em> (moving up or down the tree). In most analysis software that performs computations along trees, some other representation is used.</p>
</section>
<section id="nexus-more-soon" class="level1">
<h1>NEXUS – More soon</h1>
<p>The NEXUS format (<code>*.nex</code> or <code>*.nxs</code>) is widely used in bioinformatics. Many phylogenetics programs use nexus format including <span class="math inline">\(PAUP^*\)</span>, MrBayes, Mesquite, MacClade, IQTREE2, etc., and many more can read nexus.</p>
<p>One of the key features is that it is very <em>extensible</em> with information in <strong>blocks</strong>. We will only be concerned with blocks of DNA sequence data, but any type of information can be added. This flexibilityy, however, can be a curse when you encounter new extensions to the format that your code did not anticipate. However, for the most part we will be using packages that read in nexus so you can reasonable hope that someone has dealt with the headache for you.</p>
<p>Nexus files are often used as inputs to phylogenetic programs, to hold the DNA (or other) alignments and the taxon names. In that case, they typically only have a TAXA block and a DATA block. In our IQTREE2 example we also had a partition file in nexus format. But they can also return the output of the phylogenetic reconstruction, in that case they will have a TREES block as well.</p>
<section id="nexus-syntax" class="level3">
<h3 class="anchored" data-anchor-id="nexus-syntax">Nexus syntax:</h3>
<p>A NEXUS file begins with <code>#NEXUS</code> on the first line.</p>
<p><strong>Blocks</strong> of information follow, tyically on multiple lines, enclosed by a <strong>BEGIN blockname;</strong> and <strong>END;</strong> (Key words are case-insensitive).</p>
<p>Comments are enclosed by square brackets <code>[comments]</code></p>
<p>Typical blocks:</p>
<ul>
<li><strong>TAXA</strong>: The taxa block contains the names of the taxa.</li>
<li><strong>DATA</strong>: The data block contains a data matrix (we will use DNA sequences).</li>
<li><strong>TREES</strong>: The trees block contains one or more phylogenetic trees in newick format.</li>
</ul>
<p>For example <a href="https://en.wikipedia.org/wiki/Nexus_file">From Wikipedia</a>:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co">#NEXUS</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="ex">Begin</span> TAXA<span class="kw">;</span></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>  <span class="ex">Dimensions</span> ntax=4<span class="kw">;</span></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>  <span class="ex">TaxLabels</span> SpaceDog SpaceCat SpaceOrc SpaceElf<span class="kw">;</span></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a><span class="ex">End</span><span class="kw">;</span></span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a><span class="ex">Begin</span> data<span class="kw">;</span></span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a>  <span class="ex">Dimensions</span> nchar=15<span class="kw">;</span></span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a>  <span class="ex">Format</span> datatype=dna missing=<span class="pp">?</span> gap=- matchchar=.<span class="kw">;</span></span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a>  <span class="ex">Matrix</span></span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a>    <span class="bu">[</span> When a position <span class="er">is</span> <span class="ex">a</span> <span class="st">"matchchar"</span>, it means that it is the same as the first entry at the same position. ]</span>
<span id="cb9-12"><a href="#cb9-12" aria-hidden="true" tabindex="-1"></a>    <span class="ex">SpaceDog</span>   atgctagctagctcg</span>
<span id="cb9-13"><a href="#cb9-13" aria-hidden="true" tabindex="-1"></a>    <span class="ex">SpaceCat</span>   ......<span class="pp">??</span>...-.a.</span>
<span id="cb9-14"><a href="#cb9-14" aria-hidden="true" tabindex="-1"></a>    <span class="ex">SpaceOrc</span>   ...t.......-.g. [ same as atgttagctag-tgg ]</span>
<span id="cb9-15"><a href="#cb9-15" aria-hidden="true" tabindex="-1"></a>    <span class="ex">SpaceElf</span>   ...t.......-.a.           </span>
<span id="cb9-16"><a href="#cb9-16" aria-hidden="true" tabindex="-1"></a>  <span class="kw">;</span></span>
<span id="cb9-17"><a href="#cb9-17" aria-hidden="true" tabindex="-1"></a><span class="ex">End</span><span class="kw">;</span></span>
<span id="cb9-18"><a href="#cb9-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-19"><a href="#cb9-19" aria-hidden="true" tabindex="-1"></a><span class="ex">BEGIN</span> TREES<span class="kw">;</span></span>
<span id="cb9-20"><a href="#cb9-20" aria-hidden="true" tabindex="-1"></a>  <span class="ex">Tree</span> tree1 = <span class="er">(</span><span class="kw">((</span><span class="ex">SpaceDog,SpaceCat</span><span class="kw">)</span><span class="ex">,SpaceOrc,SpaceElf</span><span class="kw">));</span></span>
<span id="cb9-21"><a href="#cb9-21" aria-hidden="true" tabindex="-1"></a><span class="ex">END</span><span class="kw">;</span> </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="note-partition-files-in-iqtree" class="level3">
<h3 class="anchored" data-anchor-id="note-partition-files-in-iqtree">Note: Partition files in IQTREE</h3>
<p>NEXUS files are also used for a wide variety of other purposes to supply additional information to phylogenetic reconstruction. Partition files for IQTREE use a nexus format with the block labeled <code>SETS</code>, with the information coded as <code>charset LOCUSNAME = startposition - endposition;</code> The positions being along the alignment specified in the DNA alignment file (which could also be in nexus format, phylip format, or many others.</p>
<p>I like to use the gene name to label the partitions. For example, a by-locus partition file can be indicated as so:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co">#nexus</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="ex">begin</span> sets<span class="kw">;</span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>  <span class="ex">charset</span> sia = 1-407<span class="kw">;</span>  </span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>  <span class="ex">charset</span> bdnf = 408-924<span class="kw">;</span></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>  <span class="ex">charset</span> nxc = 925-1479<span class="kw">;</span></span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>  <span class="ex">charset</span> cyb = 1480-1989<span class="kw">;</span></span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>  <span class="ex">charset</span> nd4 = 1990-2475<span class="kw">;</span></span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a><span class="ex">end</span><span class="kw">;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>And a by locus and codon partition file, where the range of positions is <code>\3</code> for each third position, and the start is adjusted to indicate 1st, 2nd, 3rd codon positions:</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co">#nexus</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="ex">begin</span> sets<span class="kw">;</span></span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>  <span class="ex">charset</span> sia_pos1 = 1-407<span class="dt">\3</span><span class="kw">;</span></span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>  <span class="ex">charset</span> sia_pos2 = 2-407<span class="dt">\3</span><span class="kw">;</span></span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>  <span class="ex">charset</span> sia_pos3 = 3-407<span class="dt">\3</span><span class="kw">;</span></span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a>  <span class="ex">charset</span> bdnf_pos1 = 408-924<span class="dt">\3</span><span class="kw">;</span></span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a>  <span class="ex">charset</span> bdnf_pos2 = 409-924<span class="dt">\3</span><span class="kw">;</span></span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a>  <span class="ex">charset</span> bdnf_pos3 = 410-924<span class="dt">\3</span><span class="kw">;</span></span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a>  <span class="ex">charset</span> nxc_pos1 = 925-1479<span class="dt">\3</span><span class="kw">;</span></span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true" tabindex="-1"></a>  <span class="ex">charset</span> nxc_pos2 = 926-1479<span class="dt">\3</span><span class="kw">;</span></span>
<span id="cb11-11"><a href="#cb11-11" aria-hidden="true" tabindex="-1"></a>  <span class="ex">charset</span> nxc_pos3 = 927-1479<span class="dt">\3</span><span class="kw">;</span></span>
<span id="cb11-12"><a href="#cb11-12" aria-hidden="true" tabindex="-1"></a>  <span class="ex">charset</span> cyb_pos1 = 1480-1989<span class="dt">\3</span><span class="kw">;</span></span>
<span id="cb11-13"><a href="#cb11-13" aria-hidden="true" tabindex="-1"></a>  <span class="ex">charset</span> cyb_pos2 = 1481-1989<span class="dt">\3</span><span class="kw">;</span></span>
<span id="cb11-14"><a href="#cb11-14" aria-hidden="true" tabindex="-1"></a>  <span class="ex">charset</span> cyb_pos3 = 1482-1989<span class="dt">\3</span><span class="kw">;</span></span>
<span id="cb11-15"><a href="#cb11-15" aria-hidden="true" tabindex="-1"></a>  <span class="ex">charset</span> nd4_pos1 = 1990-2475<span class="dt">\3</span><span class="kw">;</span></span>
<span id="cb11-16"><a href="#cb11-16" aria-hidden="true" tabindex="-1"></a>  <span class="ex">charset</span> nd4_pos2 = 1991-2475<span class="dt">\3</span><span class="kw">;</span></span>
<span id="cb11-17"><a href="#cb11-17" aria-hidden="true" tabindex="-1"></a>  <span class="ex">charset</span> nd4_pos3 = 1992-2475<span class="dt">\3</span><span class="kw">;</span></span>
<span id="cb11-18"><a href="#cb11-18" aria-hidden="true" tabindex="-1"></a><span class="ex">end</span><span class="kw">;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
</section>
<section id="phylo-ape-5.0-or-above" class="level1">
<h1><code>phylo</code> (<code>ape</code> 5.0 or above)</h1>
<p>The <a href="http://ape-package.ird.fr"><code>ape</code> package</a> <span class="citation" data-cites="Paradis:2019">(<a href="#ref-Paradis:2019" role="doc-biblioref">Paradis and Schliep 2019</a>)</span> uses a different representation of trees. It uses R <code>structures</code>, <code>lists</code>, <code>matrices</code>, and <code>vectors</code> to store a tree. Each node in the tree receives a number. For example, here is the tree from before in <code>ape</code> format.</p>
<p>Here is <code>simpletree</code> with the node numbers printed. It is printed with the following commands:</p>
<div class="cell" data-caption="A simple tree with ape's numbering of nodes included">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(simpletree,<span class="at">no.margin=</span><span class="cn">TRUE</span>)</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="fu">nodelabels</span>()</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a><span class="fu">tiplabels</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="index_files/figure-html/simpletreeplotnodelabels-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>For a tree with <strong>N tips</strong>, the tips have <strong>numbers 1…N</strong> and the <strong>nodes</strong> have <strong>numbers greater than N</strong> (this is in contrast to how this was done in early (&lt;1.9) versions of <code>ape</code>). These numbers are used to store information about the tree’s structure.</p>
<p>To do this, a matrix is created, with height corresponding to the number of internal and terminal nodes and width 2. The first column of the matrix has the node at the beginning of the branch, the second has the node at the end of the branch. For example, for our simple tree, this matrix is</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>simpletree<span class="sc">$</span>edge</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      [,1] [,2]
 [1,]    8    9
 [2,]    9   10
 [3,]   10    1
 [4,]   10    2
 [5,]    9   11
 [6,]   11    3
 [7,]   11    4
 [8,]    8   12
 [9,]   12    5
[10,]   12   13
[11,]   13    6
[12,]   13    7</code></pre>
</div>
</div>
<p>This alone is enough for a basic topology. However, it might be nice to know what the taxa actually are, rather than just numbers. To do this, a character vector with as many entries as the number of tips is used. In the example tree, this is</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>simpletree<span class="sc">$</span>tip.label</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "A" "B" "C" "D" "E" "F" "G"</code></pre>
</div>
</div>
<p>It’s possible that internal nodes have labels, too (for example, the most recent common ancestor of a set of birds might be labeled Aves). If so, an optional <code>node.label</code> is used. If branch lengths are known, they are included as the numeric vector <code>edge.length</code>. Letʻs some add using ape functions:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>simpletreebrlen<span class="ot">&lt;-</span><span class="fu">compute.brlen</span>(simpletree)</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>simpletreebrlen<span class="sc">$</span>edge.length<span class="ot">&lt;-</span><span class="dv">3</span><span class="sc">*</span>simpletreebrlen<span class="sc">$</span>edge.length</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>simpletree<span class="ot">&lt;-</span>simpletreebrlen</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>simpletree<span class="sc">$</span>edge.length</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] 1.5 1.0 0.5 0.5 1.0 0.5 0.5 2.0 1.0 0.5 0.5 0.5</code></pre>
</div>
</div>
<p>Finally, there are a few other elements (<code>Nnode</code>, the number of internal nodes; <code>class=phylo</code>) to set the class. To dump the the internal representation of phylo tree to screen, you can use <code>unclass()</code> which strips the class attribute (the S4 analog is <code>attributes</code>):</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="fu">unclass</span>(simpletree)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$edge
      [,1] [,2]
 [1,]    8    9
 [2,]    9   10
 [3,]   10    1
 [4,]   10    2
 [5,]    9   11
 [6,]   11    3
 [7,]   11    4
 [8,]    8   12
 [9,]   12    5
[10,]   12   13
[11,]   13    6
[12,]   13    7

$tip.label
[1] "A" "B" "C" "D" "E" "F" "G"

$Nnode
[1] 6

$edge.length
 [1] 1.5 1.0 0.5 0.5 1.0 0.5 0.5 2.0 1.0 0.5 0.5 0.5</code></pre>
</div>
</div>
<p><code>phylo</code> trees are S3 objects. We’ll be learning more about them later, but an important thing to know is that you directly access any element of them by using the <code>$</code> operator (as was done above). Optional elements, or even elements of your own devising, can be added to them, too, using the same operator.</p>
<section id="getting-trees-into-phylo-format" class="level3">
<h3 class="anchored" data-anchor-id="getting-trees-into-phylo-format">Getting trees into phylo format</h3>
<p><code>ape</code> has many of functions for reading in trees (the <code>read.</code> series) and writing trees to file (the <code>write.</code> series). The most commonly used are:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="fu">read.nexus</span>(file, <span class="at">tree.names =</span> <span class="cn">NULL</span>, <span class="at">force.multi =</span> <span class="cn">FALSE</span>)</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a><span class="fu">read.tree</span>(<span class="at">file =</span> <span class="st">""</span>, <span class="at">text =</span> <span class="cn">NULL</span>, <span class="at">tree.names =</span> <span class="cn">NULL</span>, <span class="at">skip =</span> <span class="dv">0</span>,</span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">comment.char =</span> <span class="st">""</span>, <span class="at">keep.multi =</span> <span class="cn">FALSE</span>, ...)</span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a><span class="fu">read.FASTA</span>(file, <span class="at">type =</span> <span class="st">"DNA"</span>)</span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a><span class="fu">read.GenBank</span>(access.nb, <span class="at">seq.names =</span> access.nb, <span class="at">species.names =</span> <span class="cn">TRUE</span>,</span>
<span id="cb22-6"><a href="#cb22-6" aria-hidden="true" tabindex="-1"></a>             <span class="at">as.character =</span> <span class="cn">FALSE</span>, <span class="at">chunk.size =</span> <span class="dv">400</span>, <span class="at">quiet =</span> <span class="cn">TRUE</span>)    </span>
<span id="cb22-7"><a href="#cb22-7" aria-hidden="true" tabindex="-1"></a><span class="fu">write.nexus</span>(..., <span class="at">file =</span> <span class="st">""</span>, <span class="at">translate =</span> <span class="cn">TRUE</span>)</span>
<span id="cb22-8"><a href="#cb22-8" aria-hidden="true" tabindex="-1"></a><span class="fu">write.tree</span>(tree, <span class="at">file =</span> <span class="st">""</span>, <span class="at">append =</span> <span class="cn">FALSE</span>,</span>
<span id="cb22-9"><a href="#cb22-9" aria-hidden="true" tabindex="-1"></a>           <span class="at">digits =</span> <span class="dv">10</span>, <span class="at">tree.names =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>To see more, invoke the package help page <code>?ape</code> and look under <code>read.</code> and <code>write.</code></p>
<p>Examples:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="fu">write.tree</span>(simpletree, <span class="st">"mytree.tree"</span>)</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>tree <span class="ot">&lt;-</span> <span class="fu">read.tree</span>(<span class="st">"mytree.tree"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>tree2 <span class="ot">&lt;-</span> <span class="fu">read.tree</span>(<span class="at">text=</span><span class="st">"((G:1.0,F:1.0):1.1,E:2.1):3.5;"</span>)</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(tree2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="index_files/figure-html/unnamed-chunk-10-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p><code>phylo</code> is widely used in other packages as well. <code>ape</code> itself has many functions as it was written primarily to perform phylogenetic comparative analyses.</p>
<section id="methods" class="level4">
<h4 class="anchored" data-anchor-id="methods">Methods</h4>
<p>In line with the style of object-oriented programming, R has a generic function convention that will dispatch the correct function after looking at the class of the first object passed to it as an argument. For example, when you tell it to <code>plot(vector_object)</code> R takes a look – “oh, youʻre a vector, Iʻll send you to the plotting method for vectors”.</p>
<p>The developer of the class (usually bundled in a package) has to write all of the generic methods for their new class. You can see all of the methods available for S3 classes like <code>phylo</code> like so:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="fu">methods</span>(<span class="at">class=</span><span class="fu">class</span>(simpletree)) <span class="co"># same as methods(class="phylo")</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in findGeneric(nm, envir = as.environment(w)): 'drop.tip' is a formal
generic function; S3 methods will not likely be found</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] +                    all.equal            ancestor            
 [4] as_tibble            as.evonet            as.hclust           
 [7] as.matching          as.treedata          as.ultrametric      
[10] c                    child                coalescent.intervals
[13] coerce               cophenetic           degree              
[16] di2multi             drop.tip             full_join           
[19] identify             initialize           is.binary           
[22] is.rooted            is.ultrametric       isTip               
[25] keep.tip             makeLabel            MRCA                
[28] multi2di             Nedge                Nnode               
[31] nodeid               nodelab              Ntip                
[34] offspring            parent               plot                
[37] print                reorder              root                
[40] rootnode             show                 skyline             
[43] slotsFromS3          summary              tree_subset         
[46] unroot               updateLabel          vcv                 
see '?methods' for accessing help and source code</code></pre>
</div>
</div>
<p>Remember these are functions, so these are all functions that are written to work with any <code>phylo</code> object.</p>
<p>Try some, like <code>Ntip(simpletree)</code>. Both <code>drop.tip</code> and <code>keep.tip</code> are really helpful for cutting your tree down to a subset of taxa.</p>
</section>
</section>
</section>
<section id="ouchtree-ouch-version-2.0-or-above" class="level1">
<h1>ouchtree (<code>ouch</code> version 2.0 or above)</h1>
<p>Trees in the package <code>ouch</code> <span class="citation" data-cites="Butler:2004">King and Butler (<a href="#ref-King:2022" role="doc-biblioref">2022</a>)</span> are of class <code>ouchtree</code>, and are written as an S4 object, as opposed to the S3 object class used by <code>ape</code>.</p>
<p>Whereas S3 objects can be created on the fly by putting primitive object types together, S4 objects are created by a constructor function that puts the object together in a more structured way and importantly, <em>checks that the object is actually constructed correctly</em>. This makes it a lot easier for developers to reliably depend on objects of that class.</p>
<section id="building-ouchtree-format-from-dataframes" class="level3">
<h3 class="anchored" data-anchor-id="building-ouchtree-format-from-dataframes">Building ouchtree format from dataframes</h3>
<p>The <code>ouchtree()</code> function is the constructor for objects of class <code>ouchtree</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ouchtree</span>(nodes, ancestors, times, <span class="at">labels =</span> <span class="fu">as.character</span>(nodes))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><strong>The <code>ouchtree</code> function only has four arguments</strong>, one of them optional.</p>
<ul>
<li><code>nodes</code>: a character vector of node ids (including internal nodes). Unlike <code>ape</code>, the leaves do not need to have smaller ids than internal nodes.</li>
<li><code>ancestors</code>: a character vector of node ids of the ancestors for the nodes in the <code>nodes</code> vector. The <code>nodes</code> and <code>ancestors</code> vectors almost correspond to the second and first columns of the <code>ape</code> <code>edge</code> matrix, respectively, with the exception that <code>ouchtree</code> includes the root node with an ancestor of <code>NA</code>.</li>
<li><code>times</code>: represents the height of each node from the root. Remember that <code>ape</code>’s <code>edge.length</code> vector has the length of the branch subtending each node; instead, <code>ouchtree</code> has the sum of the lengths of all branches connecting a given node to the root. Again, the root node is included in <code>ouchtree</code> (with height 0) but not in ape.</li>
<li><code>labels</code>: is a vector of labels for both tips and internal nodes. It is optional, if no argument is supplied then the <code>nodes</code> are used as labels. If internal nodes do not have names, they get a label of <code>&lt;NA&gt;</code>.</li>
</ul>
<p>We can use <code>ouch</code>ʻs built-in dataset <code>anolis.ssd</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(anolis.ssd)</span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(anolis.ssd)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "data.frame"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>anolis.ssd</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   node species  log.SSD ancestor   time OU.1 OU.7
1     1    &lt;NA&gt;       NA       NA 0.0000   ns  anc
2     2    &lt;NA&gt;       NA        1 0.1614   ns  anc
3     3    &lt;NA&gt;       NA        2 0.1845   ns  anc
4     4    &lt;NA&gt;       NA        3 0.4778   ns  anc
5     5    &lt;NA&gt;       NA        3 0.3909   ns  anc
6     6    &lt;NA&gt;       NA        3 0.5026   ns  anc
7     7    &lt;NA&gt;       NA        3 0.2372   ns  anc
8     8    &lt;NA&gt;       NA        7 0.6986   ns  anc
9     9    &lt;NA&gt;       NA        7 0.4864   ns  anc
10   10    &lt;NA&gt;       NA        9 0.6838   ns  anc
11   11    &lt;NA&gt;       NA        3 0.2783   ns  anc
12   12    &lt;NA&gt;       NA       11 0.5529   ns  anc
13   13    &lt;NA&gt;       NA       11 0.4259   ns  anc
14   14    &lt;NA&gt;       NA        3 0.2826   ns  anc
15   15    &lt;NA&gt;       NA       14 0.3998   ns  anc
16   16      oc -0.00512        1 1.0000   ns   tw
17   17      eq  0.08454        2 1.0000   ns   cg
18   18      co  0.24703        4 1.0000   ns   tc
19   19     aln  0.24837        4 1.0000   ns   tc
20   20      ol  0.09844        5 1.0000   ns   gb
21   21      in  0.06137        5 1.0000   ns   tw
22   22      cu  0.09501        6 1.0000   ns   cg
23   23      ri  0.08947        6 1.0000   ns   cg
24   24      sa  0.28893        8 1.0000   ns   tg
25   25      op  0.15963        8 1.0000   ns   gb
26   26      va  0.14766        9 1.0000   ns   tw
27   27      li  0.38873        9 1.0000   ns   tg
28   28      ga  0.28768       10 1.0000   ns   cg
29   29      gr  0.39786       10 1.0000   ns   tc
30   30      br  0.17106       12 1.0000   ns   tr
31   31      di  0.13801       12 1.0000   ns   tr
32   32      kr  0.23478       13 1.0000   ns   gb
33   33      cr  0.39796       13 1.0000   ns   tg
34   34      st  0.15737       13 1.0000   ns   tc
35   35      cy  0.26024        3 1.0000   ns   tg
36   36     alu  0.08216       14 1.0000   ns   gb
37   37      lo  0.13108       15 1.0000   ns   tr
38   38      an  0.13547       15 1.0000   ns   tw</code></pre>
</div>
</div>
<p>This is a dataframe containing sexual size dimorphism data on 38 species of <code>Anolis</code> lizard <span class="citation" data-cites="Butler:2004">(<a href="#ref-Butler:2004" role="doc-biblioref">Butler and King 2004</a>)</span>. The tree information is provided by <code>node</code>, <code>species</code>, <code>ancestor</code>, and <code>time</code>. The <code>log.SSD</code> contains the phenotypic data we are interested in studying, and the other vectors (<code>OU.1</code>, <code>OU.7</code>) contain hypotheses for the evolution of these phenotypes (they are potential covariates). (More on this later).</p>
<p>We construct the ouchtree like so, and check with a plot:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>tree <span class="ot">&lt;-</span> <span class="fu">with</span>(anolis.ssd,<span class="fu">ouchtree</span>(node,ancestor,time<span class="sc">/</span><span class="fu">max</span>(time),species))</span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(tree)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="index_files/figure-html/unnamed-chunk-14-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="converting-from-ape-to-ouch" class="level3">
<h3 class="anchored" data-anchor-id="converting-from-ape-to-ouch">Converting from <code>ape</code> to <code>ouch</code></h3>
<p>However, most people donʻt know how to easily convert their newick or nexus format trees to a dataframe (and would be afraid to make errors), so probably the most commonly used method to obtain ouchtrees is to read in the treefile to <code>phylo</code> format using <code>ape</code>, then use ouchʻs <code>ape2ouch()</code> function.</p>
<p>For example:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>osimpletree<span class="ot">&lt;-</span><span class="fu">ape2ouch</span>(<span class="at">tree=</span>simpletree)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Our example tree, when converted to ouchtree format, is</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>osimpletree</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   nodes ancestors     times labels
1      1      &lt;NA&gt; 0.0000000       
2      2         3 0.8333333       
3      3         1 0.6666667       
4      4         6 0.8333333       
5      5         6 0.8333333       
6      6         1 0.5000000       
7      7         2 1.0000000      G
8      8         2 1.0000000      F
9      9         3 1.0000000      E
10    10         4 1.0000000      D
11    11         4 1.0000000      C
12    12         5 1.0000000      B
13    13         5 1.0000000      A</code></pre>
</div>
</div>
<p>Note that the print method for the tree displays the vectors used to construct the tree (the user input).</p>
</section>
<section id="coerce-to-data.frame" class="level3">
<h3 class="anchored" data-anchor-id="coerce-to-data.frame">Coerce to data.frame</h3>
<p>ouchtree objects can also be coerced to dataframes, which is helpful for adding hypotheses or metadata. Because <code>ouchtree</code> is an S4 object, to coerce it to a dataframe you use:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="fu">as</span>(osimpletree, <span class="st">"data.frame"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   nodes ancestors     times labels
1      1      &lt;NA&gt; 0.0000000       
2      2         3 0.8333333       
3      3         1 0.6666667       
4      4         6 0.8333333       
5      5         6 0.8333333       
6      6         1 0.5000000       
7      7         2 1.0000000      G
8      8         2 1.0000000      F
9      9         3 1.0000000      E
10    10         4 1.0000000      D
11    11         4 1.0000000      C
12    12         5 1.0000000      B
13    13         5 1.0000000      A</code></pre>
</div>
</div>
</section>
<section id="hidden-components-of-ouchtree" class="level3">
<h3 class="anchored" data-anchor-id="hidden-components-of-ouchtree">Hidden components of <code>ouchtree</code></h3>
<p><code>ouch</code> was written primarily to perform computations along phylogenies to test models of evolution, and thus the <code>ouctree</code> object constructor creates many informational matrices and vectors that are used for rapid computations. Itʻs plotting function is mainly to check that the phylogeny and models were entered correctly, but itʻs also probably the fastest plotting algorithm around.</p>
<p>For example, the <code>branch.times</code> matrix, created on initialization, is part of the <code>ouchtree</code> object and is a matrix showing shared amount of time on a tree between two tips (which may be the same tip). This, multiplied by a rate parameter, becomes a variance-covariance matrix under a Brownian motion model, which we’ll be discussing later.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="fu">attributes</span>(osimpletree)<span class="sc">$</span>branch.times</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>          [,1]      [,2]      [,3]      [,4]      [,5]      [,6]      [,7]
[1,] 1.0000000 0.8333333 0.6666667 0.0000000 0.0000000 0.0000000 0.0000000
[2,] 0.8333333 1.0000000 0.6666667 0.0000000 0.0000000 0.0000000 0.0000000
[3,] 0.6666667 0.6666667 1.0000000 0.0000000 0.0000000 0.0000000 0.0000000
[4,] 0.0000000 0.0000000 0.0000000 1.0000000 0.8333333 0.5000000 0.5000000
[5,] 0.0000000 0.0000000 0.0000000 0.8333333 1.0000000 0.5000000 0.5000000
[6,] 0.0000000 0.0000000 0.0000000 0.5000000 0.5000000 1.0000000 0.8333333
[7,] 0.0000000 0.0000000 0.0000000 0.5000000 0.5000000 0.8333333 1.0000000</code></pre>
</div>
</div>
<section id="ouchtree-slots" class="level4">
<h4 class="anchored" data-anchor-id="ouchtree-slots"><code>ouchtree</code> Slots</h4>
<p>The <code>ouchtree()</code> function (from the 3-4 user supplied arguments), initializes all of the other elements (slots) of class <code>ouchtree</code> which is defined as:</p>
<ul>
<li><code>nnodes</code> = ‘integer’ vector, number of nodes</li>
<li><code>nodes</code> = ‘character’ vector, names for nodes</li>
<li><code>ancestors</code> = ‘character’, names for ancestors</li>
<li><code>nodelabels</code> = ‘character’, user-specified labels (i.e., species names)</li>
<li><code>times</code> = ‘numeric’, time from root to node</li>
<li><code>root</code> = ‘integer’, identity of the root node</li>
<li><code>nterm</code> = ‘integer’, number of terminal taxa or tips</li>
<li><code>term</code> = ‘integer’, vector of terminal nodes</li>
<li><code>anc.numbers</code> = ‘integer’, numbers of ancestors</li>
<li><code>lineages</code> = ‘list’, nodes traversed from root to each node</li>
<li><code>epochs</code> = ‘list’,</li>
<li><code>branch.times</code> = ‘matrix’,</li>
<li><code>depth</code> = ‘numeric’</li>
</ul>
<p>The entire content of the <code>simpletreeouch</code> object can be dumped to screen using the following command (not executed here to save paper):</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="fu">attributes</span>(simpletreeouch)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We will return to ouch in the comparative methods section.</p>
</section>
</section>
</section>
<section id="treedata-from-package-tidytree" class="level1">
<h1><code>treedata</code> from package <code>tidytree</code></h1>
<p>The class <code>treedata</code> comes from the <code>tidytree</code> package which was developed by Guangchuang Yu and colleagues <span class="citation" data-cites="Yu:2022">(<a href="#ref-Yu:2022" role="doc-biblioref">Yu 2022</a>)</span>, who also developed the <code>treeio</code> package which is the <em>Swiss Army Knife</em> of phylogenetic tree formats (<a href="#fig-treeio">Figure&nbsp;2</a>), and the <code>ggtree</code> package which is quickly becoming the tree plotting package of choice. <code>ggtree</code> was written to use <code>ggplot2</code> and in general all three packages interact nicely with the <code>tidyverse</code>.</p>
<div id="fig-treeio" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="../../images/treeio-diagram.png" class="img-fluid figure-img"></p>
<p></p><figcaption class="figure-caption">Figure&nbsp;2: Overview of the treeio package and its relations with tidytree and ggtree. Treeio supports parsing a tree with data from a number of file formats and software outputs. A treedata object stores a phylogenetic tree with node/branch-associated data. Treeio provides several functions to manipulate a tree with data. Users can convert the treedata object into a tidy data frame (each row represents a node in the tree and each column represents a variable) and process the tree with data using the tidy interface implemented in tidytree. The tree can be extracted from the treedata object and exported to a Newick and NEXUS file or can be exported with associated data into a single file (either in the BEAST NEXUS or jtree format). Associated data stored in the treedata object can be used to annotate the tree using ggtree. In addition, ggtree supports a number of tree objects, including phyloseq for microbiome data and obkData for outbreak data. The phylo, multiPhylo (ape package), phylo4, phylo4d (phylobase package), phylog (ade4 package), phyloseq (phyloseq package), and obkData (OutbreakTools package) are tree objects defined by the R community to store tree with or without domain-specific data. All these tree objects as well as hierarchical clustering results (e.g., hclust and dendrogram objects) are supported by ggtree.</figcaption><p></p>
</figure>
</div>
<p><a href="https://yulab-smu.top/treedata-book/chapter1.html">Source: Yu, 2022 Fig. 1.3</a></p>
<p>Extensive examples and explanation are provided in his new book, which is availble online <a href="https://yulab-smu.top/treedata-book" class="uri">https://yulab-smu.top/treedata-book</a>. To get started on <a href="https://yulab-smu.top/treedata-book/chapter2.html#the-treedata-object">treedata check out this page</a>.</p>
<p><code>treedata</code> is also an S4 object, which is excellent, as it is easy to hand-create a <code>phylo</code> object with some content that may be not obvious to the user but will cause functions to throw errors. S4 objects avoid all that by having constructor functions.</p>
<p>The philosophy of <code>treedata</code> is to create rigorous tree objects that store phylogenetic trees with associated data. One main drawback of class <code>phylo</code> is that trees and data are stored separately, and it is up to the user to ensure that the data lines up with <code>phylo</code>ʻs edge matrix properly. This is scary! <code>treedata</code> allow the opportunity to bind the tree and the data together correctly.</p>
<section id="treedata-slots" class="level4">
<h4 class="anchored" data-anchor-id="treedata-slots"><code>treedata</code> Slots</h4>
<p>Class <code>treedata</code> stores phylogenetic tree with associated data</p>
<ul>
<li><code>file</code> = tree file</li>
<li><code>treetext</code> = newick tree string</li>
<li><code>phylo</code> = phylo object for tree structure</li>
<li><code>data</code> = associated data</li>
<li><code>extraInfo</code> = extra information, reserve for merge_tree</li>
<li><code>tip_seq</code> = tip sequences</li>
<li><code>anc_seq</code> = ancestral sequences</li>
<li><code>seq_type</code> = sequence type, one of NT or AA</li>
<li><code>tipseq_file</code> = tip sequence file</li>
<li><code>ancseq_file</code> = ancestral sequence file</li>
<li><code>info</code> = extra information, e.g.&nbsp;metadata, software version etc.</li>
</ul>
</section>
<section id="reading-trees-into-treedata" class="level3">
<h3 class="anchored" data-anchor-id="reading-trees-into-treedata">Reading trees into <code>treedata</code></h3>
<p>There are functions to read in tree formats from every major phylogenetics reconstruction formats. There are <code>read.</code>, functions for all of the formats illustrated in <a href="#fig-treeio">Figure&nbsp;2</a>, such as <code>read.nexus</code>, <code>read.beast</code> and <code>read.iqtree</code>.</p>
<p>It is also easy to coerce <code>phylo</code> to <code>treedata</code> or vice versa:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>tdat <span class="ot">&lt;-</span> <span class="fu">as.treedata</span>(simpletree)</span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a>tdat</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>'treedata' S4 object'.

...@ phylo:

Phylogenetic tree with 7 tips and 6 internal nodes.

Tip labels:
  A, B, C, D, E, F, ...

Rooted; includes branch lengths.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="fu">as.phylo</span>(tdat)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Phylogenetic tree with 7 tips and 6 internal nodes.

Tip labels:
  A, B, C, D, E, F, ...

Rooted; includes branch lengths.</code></pre>
</div>
</div>
<p>A lot more detail is provided in the book sections 1.2, 1.3 <span class="citation" data-cites="Yu:2022">(<a href="#ref-Yu:2022" role="doc-biblioref">Yu 2022</a>)</span>.</p>
</section>
<section id="combining-tree-with-data" class="level3">
<h3 class="anchored" data-anchor-id="combining-tree-with-data">Combining tree with data</h3>
<p><code>treedata</code> is a special kind of tibble. So again, it is easy to undersatnd as it is presented as a rectangular dataframe-like object.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a>simpletree <span class="sc">%&gt;%</span> as.treedata <span class="sc">%&gt;%</span> as_tibble</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 13 × 4
   parent  node branch.length label
    &lt;dbl&gt; &lt;dbl&gt;         &lt;dbl&gt; &lt;chr&gt;
 1     10     1           0.5 A    
 2     10     2           0.5 B    
 3     11     3           0.5 C    
 4     11     4           0.5 D    
 5     12     5           1   E    
 6     13     6           0.5 F    
 7     13     7           0.5 G    
 8      8     8          NA   &lt;NA&gt; 
 9      8     9           1.5 &lt;NA&gt; 
10      9    10           1   &lt;NA&gt; 
11      9    11           1   &lt;NA&gt; 
12      8    12           2   &lt;NA&gt; 
13     12    13           0.5 &lt;NA&gt; </code></pre>
</div>
</div>
</section>
<section id="methods-to-extract-information-from-treedata" class="level3">
<h3 class="anchored" data-anchor-id="methods-to-extract-information-from-treedata">Methods to extract information from <code>treedata</code></h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a><span class="fu">get.tree</span>()</span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a><span class="fu">get.data</span>()</span>
<span id="cb48-3"><a href="#cb48-3" aria-hidden="true" tabindex="-1"></a><span class="fu">get.fields</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="plotting-ggtree" class="level3">




</section>
</section>

<div id="quarto-appendix" class="default"><section class="quarto-appendix-contents" role="doc-bibliography"><h2 class="anchored quarto-appendix-heading">Plotting ggtree</h2><div id="refs" class="references csl-bib-body hanging-indent" role="doc-bibliography">
<div id="ref-Butler:2004" class="csl-entry" role="doc-biblioentry">
Butler, Marguerite A., and Aaron A. King. 2004. <span>“Phylogenetic Comparative Analysis: A Modeling Approach for Adaptive Evolution.”</span> <em>American Naturalist</em> 164: 683–95. <a href="https://doi.org/10.1086/426002">https://doi.org/10.1086/426002</a>.
</div>
<div id="ref-King:2022" class="csl-entry" role="doc-biblioentry">
King, Aaron A., and Marguerite A. Butler. 2022. <em>Ouch: Ornstein-Uhlenbeck Models for Phylogenetic Comparative Hypotheses</em>. <a href="https://kingaa.github.io/ouch/">https://kingaa.github.io/ouch/</a>.
</div>
<div id="ref-Paradis:2019" class="csl-entry" role="doc-biblioentry">
Paradis, Emmanuel, and Klaus Schliep. 2019. <span>“Ape 5.0: An Environment for Modern Phylogenetics and Evolutionary Analyses in <span>R</span>.”</span> <em>Bioinformatics</em> 35: 526–28. <a href="https://doi.org/10.1093/bioinformatics/bty633">https://doi.org/10.1093/bioinformatics/bty633</a>.
</div>
<div id="ref-Yu:2022" class="csl-entry" role="doc-biblioentry">
Yu, Guangchuang. 2022. <em>Data Integration, Manipulation and Visualization of Phylogenetic Treess</em>. 1st edition. Chapman; Hall/<span>CRC</span>. <a href="https://doi.org/10.1201/9781003279242">https://doi.org/10.1201/9781003279242</a>.
</div>
</div></section></div></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>