<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.2.313">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Marguerite Butler">
<meta name="dcterms.date" content="2023-04-13">
<meta name="description" content="Maximum Likelihood phylogenetic tree inference and gene-tree species-tree concordance">

<title>ZOOL710 Data Science in R for Biologists - Phylogenetic Inference with IQTREE2</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>

  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<link rel="stylesheet" href="../../styles.css">
</head>

<body class="floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">ZOOL710 Data Science in R for Biologists</span>
    </a>
  </div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../index.html">
 <span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-general-information" role="button" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">General Information</span>
    </a>
    <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="nav-menu-general-information">    
        <li>
    <a class="dropdown-item" href="../../syllabus.html">
 <span class="dropdown-text">Syllabus</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../schedule.html">
 <span class="dropdown-text">Schedule</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-course-materials" role="button" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Course Materials</span>
    </a>
    <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="nav-menu-course-materials">    
        <li>
    <a class="dropdown-item" href="../../lectures.html">
 <span class="dropdown-text">Lectures</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../projects.html">
 <span class="dropdown-text">Projects</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../resources.html">
 <span class="dropdown-text">Resources</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/"><i class="bi bi-github" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://twitter.com"><i class="bi bi-twitter" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="../../index.xml"><i class="bi bi-rss" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
</ul>
              <div id="quarto-search" class="" title="Search"></div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<header id="title-block-header" class="quarto-title-block default toc-left page-columns page-full">
  <div class="quarto-title-banner page-columns page-full">
    <div class="quarto-title column-body">
      <h1 class="title">Phylogenetic Inference with IQTREE2</h1>
                  <div>
        <div class="description">
          Maximum Likelihood phylogenetic tree inference and gene-tree species-tree concordance
        </div>
      </div>
                          <div class="quarto-categories">
                <div class="quarto-category">module 7</div>
                <div class="quarto-category">week 13</div>
                <div class="quarto-category">iqtree2</div>
                <div class="quarto-category">phylogenetic inference</div>
                <div class="quarto-category">command line</div>
                <div class="quarto-category">gene concordance</div>
                <div class="quarto-category">site concordance</div>
                <div class="quarto-category">gene tree</div>
                <div class="quarto-category">species tree</div>
              </div>
                  </div>
  </div>
    
  <div class="quarto-title-meta-author">
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-heading">Affiliation</div>
    
      <div class="quarto-title-meta-contents">
      <a href="https://butlerlab.org">Marguerite Butler</a> 
    </div>
      <div class="quarto-title-meta-contents">
          <p class="affiliation">
              School of Life Sciences, University of Hawaii
            </p>
        </div>
      </div>

  <div class="quarto-title-meta">

        
      <div>
      <div class="quarto-title-meta-heading">Published</div>
      <div class="quarto-title-meta-contents">
        <p class="date">April 13, 2023</p>
      </div>
    </div>
    
      
    </div>
    
  
  </header><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse sidebar-navigation floating overflow-auto">
    <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#acknowledgements" id="toc-acknowledgements" class="nav-link active" data-scroll-target="#acknowledgements">Acknowledgements</a></li>
  <li><a href="#learning-objectives" id="toc-learning-objectives" class="nav-link" data-scroll-target="#learning-objectives">Learning objectives</a></li>
  <li><a href="#overview" id="toc-overview" class="nav-link" data-scroll-target="#overview">Overview</a></li>
  <li><a href="#iqtree-and-ml-phylogenetic-inference-lecture" id="toc-iqtree-and-ml-phylogenetic-inference-lecture" class="nav-link" data-scroll-target="#iqtree-and-ml-phylogenetic-inference-lecture">IQTREE and ML phylogenetic inference lecture</a></li>
  <li><a href="#infer-the-turtle-phylogeny-with-iqtree" id="toc-infer-the-turtle-phylogeny-with-iqtree" class="nav-link" data-scroll-target="#infer-the-turtle-phylogeny-with-iqtree">Infer the Turtle Phylogeny with IQTREE</a>
  <ul class="collapse">
  <li><a href="#the-theoretical-problem" id="toc-the-theoretical-problem" class="nav-link" data-scroll-target="#the-theoretical-problem">The theoretical problem</a></li>
  <li><a href="#the-data" id="toc-the-data" class="nav-link" data-scroll-target="#the-data">The data</a></li>
  <li><a href="#species-trees" id="toc-species-trees" class="nav-link" data-scroll-target="#species-trees">Species Trees</a>
  <ul class="collapse">
  <li><a href="#simplest-phylogeny" id="toc-simplest-phylogeny" class="nav-link" data-scroll-target="#simplest-phylogeny">Simplest phylogeny</a></li>
  <li><a href="#species-tree-partitioned-by-locus" id="toc-species-tree-partitioned-by-locus" class="nav-link" data-scroll-target="#species-tree-partitioned-by-locus">Species Tree partitioned by locus</a></li>
  <li><a href="#gene-trees" id="toc-gene-trees" class="nav-link" data-scroll-target="#gene-trees">Gene Trees</a></li>
  </ul></li>
  <li><a href="#concordance-factors" id="toc-concordance-factors" class="nav-link" data-scroll-target="#concordance-factors">Concordance factors</a></li>
  </ul></li>
  <li><a href="#further-analyses-of-iqtree-outputs" id="toc-further-analyses-of-iqtree-outputs" class="nav-link" data-scroll-target="#further-analyses-of-iqtree-outputs">Further analyses of IQTREE outputs</a>
  <ul class="collapse">
  <li><a href="#using-concordance-factors-to-understand-your-data" id="toc-using-concordance-factors-to-understand-your-data" class="nav-link" data-scroll-target="#using-concordance-factors-to-understand-your-data">Using concordance factors to understand your data</a></li>
  </ul></li>
  <li><a href="#how-concordance-factors-relate-to-each-other-and-to-bootstraps" id="toc-how-concordance-factors-relate-to-each-other-and-to-bootstraps" class="nav-link" data-scroll-target="#how-concordance-factors-relate-to-each-other-and-to-bootstraps">How concordance factors relate to each other and to bootstraps</a></li>
  <li><a href="#digging-deeper-using-discordance-factors" id="toc-digging-deeper-using-discordance-factors" class="nav-link" data-scroll-target="#digging-deeper-using-discordance-factors">Digging deeper using discordance factors</a></li>
  <li><a href="#using-concordance-factors-to-test-the-assumptions-of-an-ils-model" id="toc-using-concordance-factors-to-test-the-assumptions-of-an-ils-model" class="nav-link" data-scroll-target="#using-concordance-factors-to-test-the-assumptions-of-an-ils-model">Using concordance factors to test the assumptions of an ILS model</a></li>
  <li><a href="#internode-certainty" id="toc-internode-certainty" class="nav-link" data-scroll-target="#internode-certainty">Internode certainty</a></li>
  <li><a href="#iqtree-documentation" id="toc-iqtree-documentation" class="nav-link" data-scroll-target="#iqtree-documentation">IQTREE documentation</a></li>
  </ul>
</nav>
</nav>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
    </div>
<!-- main -->
<main class="content quarto-banner-title-block" id="quarto-document-content">




<section id="acknowledgements" class="level3">
<h3 class="anchored" data-anchor-id="acknowledgements">Acknowledgements</h3>
<p>References for this Material:</p>
<ul>
<li><p>The Gene Tree species tree tutorial by Bui Minh <a href="http://www.iqtree.org/workshop/molevol2022" class="uri">http://www.iqtree.org/workshop/molevol2022</a>. Please see for more information.</p></li>
<li><p>And Rob Lanfearʻs excellent blog post about concordance factors <a href="http://www.robertlanfear.com/blog/files/concordance_factors.html" class="uri">http://www.robertlanfear.com/blog/files/concordance_factors.html</a> with R code to help with analysis.</p></li>
<li><p>See Methods Paper: Bui Quang Minh, Matthew W Hahn, Robert Lanfear New Methods to Calculate Concordance Factors for Phylogenomic Datasets Molecular Biology and Evolution, Volume 37, Issue 9, September 2020, Pages 2727–2733, <a href="https://doi.org/10.1093/molbev/msaa106" class="uri">https://doi.org/10.1093/molbev/msaa106</a></p></li>
</ul>
</section>
<section id="learning-objectives" class="level1">
<h1>Learning objectives</h1>
<div class="callout-note callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Learning objectives
</div>
</div>
<div class="callout-body-container callout-body">
<p><strong>At the end of this lesson you will:</strong></p>
<ul>
<li>Understand where user-installed software goes on your computer</li>
<li>Be able to install software and update the search paths to run from the command line</li>
<li>Run iqtree2 from the command line on your own computer</li>
</ul>
</div>
</div>
<div class="callout-note callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Setup iqtree software
</div>
</div>
<div class="callout-body-container callout-body">
<p>Instructions for installing IQTREE2 and setup for this lession were provided in the <a href="../../posts/2023-04-13-iqtree-setup/index.html">previous lesson</a>.</p>
<section id="setup-your-tutorial-folder" class="level4">
<h4 class="anchored" data-anchor-id="setup-your-tutorial-folder">Setup your Tutorial Folder</h4>
<p>Make a <code>TurtleTutorial</code> folder, add the following files/folders:</p>
<ul>
<li>Make an <code>input</code> folder and put these files into it:
<ul>
<li><a href="http://www.iqtree.org/workshop/data/turtle.fa">turtle.fa</a> : A sequence alignment (in FASTA format) which is a subset of the original dataset used to assess the phylogenetic position of Turtles relative to Crocodiles and Birds <a href="https://doi.org/10.1186/1741-7007-10-65">Chiari et al., 2012</a>.<br>
</li>
<li><a href="http://www.iqtree.org/workshop/data/turtle.nex">turtle.nex</a> : The partition file (in NEXUS format) defining 29 genes, which are a subset of the original dataset of 248 genes.<br>
</li>
</ul></li>
<li>The R script <a href="../../code/concordance.R"><code>concordance.R</code></a></li>
<li>iqtree shell script: <a href="../../code/iqtreerun.sh"><code>iqtreerun.sh</code></a></li>
<li>If you donʻt want to setup IQTREE2 and infer the phylogenies yourself, you can clone the repo (or clone the repo, delete the out files, and run the phylogenies yourself)<a href="https://github.com/mbutler808/TurtleTutorial" class="uri">https://github.com/mbutler808/TurtleTutorial</a></li>
</ul>
</section>
<section id="r-packages-you-will-need" class="level4">
<h4 class="anchored" data-anchor-id="r-packages-you-will-need">R packages you will need</h4>
<ul>
<li><code>viridis</code></li>
<li><code>GGally</code></li>
<li><code>entropy</code></li>
<li><code>ggplot2</code></li>
<li><code>dplyr</code></li>
<li><code>ggrepel</code></li>
</ul>
</section>
</div>
</div>
</section>
<section id="overview" class="level1">
<h1>Overview</h1>
<p>When inferring phylogenies from sequence data, it is important to provide support values, usually at the nodes, to indicate how much evidence there is for each split along the phylogeny. The most common and familar are nodal support values - <strong>bootstrap support</strong> in percent for <strong>maximum likelihood trees</strong>, and <strong>posterior probability</strong> for trees inferred by <strong>Bayesian inference</strong>.</p>
<p>IQTREE2 is software for the inference of phylogenies by <strong>Maximum Likelihood</strong> <span class="citation" data-cites="Minh:2020">(<a href="#ref-Minh:2020" role="doc-biblioref">Minh et al. 2020</a>)</span>. Several additional modules are bundled with IQTREE2. <strong>ModelFinder</strong> searches for the best substition models prior to phylogenetic inference <span class="citation" data-cites="Kalyaanamoorthy:2017">(<a href="#ref-Kalyaanamoorthy:2017" role="doc-biblioref">Kalyaanamoorthy et al. 2017</a>)</span>. ModelFinder can also combine partitions to reduce overparameterization, similar to <strong>PartitionFinder</strong> <span class="citation" data-cites="Lanfear:2016">(<a href="#ref-Lanfear:2016" role="doc-biblioref">Lanfear et al. 2016</a>)</span>.</p>
<p>There is an increasing recognition that nodal support is not enough. Most phylogenetic analyses are now multi-locus, with data from several to hundred of genes. More studies are discovering that not all of the loci may agree – sometimes certain loci will support one topology whereas others may support an alternative topology. This is called gene-tree species-tree discordance. It can arise by evolutionary mechanisms such as incomplete lineage sorting when you have rapid radiation (short branch lengths in parts of the tree), where ancestral alleles do not have sufficient time to sort themselves out amongst the descendant lineages to reflect the history of cladogenesis, or it can arise by noisy data. Either way, if you have gene-tree species-tree discordance, it will erode support at nodes, or in the worst case, appear to strongly support a spurious evolutionary scenario. We can test for concordance between gene trees and species trees <span class="citation" data-cites="Minh:2020a">(<a href="#ref-Minh:2020a" role="doc-biblioref">Minh, Hahn, and Lanfear 2020</a>)</span></p>
<p>In this module we will learn how to use <a href="www.iqtree.org">IQTREE2</a> to infer phylogenetic trees (both species trees and gene trees), and compute several concordance factors. Gene concordance and discordance factors, and site concordance and discordance factors (site = position along the DNA sequence). We will use R to read in and explore the output from IQTREE.</p>
</section>
<section id="iqtree-and-ml-phylogenetic-inference-lecture" class="level1">
<h1>IQTREE and ML phylogenetic inference lecture</h1>
<p>We will review Bui Quang Minhʻs (<a href="https://bqminh.github.io/people/minh/">Minh Bùi</a>) <a href="https://github.com/simon-ho/SydneyPhyloWorkshop/blob/master/Lecture2_1.IQTREE.pdf">lecture on IQTREE2</a> and how it addresses some issues in maximum likelihood inference of phylogenetics.</p>
<p>Here is an illustration of phylogenetic inference by Maximum Likelihood (<a href="#fig-ml">Figure&nbsp;1</a>). The likelihood of the tree is calculated given the data (here, the DNA sequence at each position or site). Each site is considered an independent data point, and each variable site provides some information about how closely related the taxa are. The likelihood is an expression of a combined probability across all of the sites, assuming a particular substitution model (the best substitution models are what ModelFinder or PartitionFinder finds). The topology and branch lengths that provide the Maximum Likelihood of the data is the ML tree.</p>
<div id="fig-ml" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="../../images/Site_pattern_frequencies_models.jpg" class="img-fluid figure-img"></p>
<p></p><figcaption class="figure-caption">Figure&nbsp;1: Figure: An illustration of phylogenetic inference by Maximum Likelihood. A multiple sequence alignment (here, DNA sequences), and DNA substitution models are used to make evolutionary inferences among the taxa (four species t1 through t4). The data in this alignment (in this case a toy example with 18 sites) is converted to a set of site patterns. The site patterns are shown along with the number of times they occur in alignment. These site patterns are used to calculate the likelihood given the substitution model and a phylogenetic tree (in this case an unrooted four-taxon tree). It is also necessary to assume a substitution model to estimate evolutionary distances for pairs of sequences (distances are the number of substitutions that have occurred since sequences had a common ancestor). The evolutionary distance equation (d12) is based on the simple model proposed by Jukes and Cantor in 1969. The equation transforms the proportion of nucleotide differences between taxa 1 and 2 (p12 = 4/18; the four site patterns that differ between taxa 1 and 2 are indicated with asterisks) into an evolutionary distance (in this case d12=0.2635 substitutions per site).</figcaption><p></p>
</figure>
</div>
<p><a href="https://commons.wikimedia.org/w/index.php?curid=95686405">Source: By EBraun68 - Own work, CC BY-SA 4.0</a></p>
</section>
<section id="infer-the-turtle-phylogeny-with-iqtree" class="level1">
<h1>Infer the Turtle Phylogeny with IQTREE</h1>
<section id="the-theoretical-problem" class="level2">
<h2 class="anchored" data-anchor-id="the-theoretical-problem">The theoretical problem</h2>
<p>The position of turtles among vertebrates has been enigmatic for decades (<a href="#fig-turtle">Figure&nbsp;2</a>). At various times it has been allied more closely with mammals or lizards. It is one of those truly hard biological problems. Even with the development of genomics, the position of turtles with respect to crocodiles and birds remains a hot topic of research as of late.</p>
<div id="fig-turtle" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="../../images/turtle.png" class="img-fluid figure-img"></p>
<p></p><figcaption class="figure-caption">Figure&nbsp;2: The changing position of turtles. Image by Jeremy Brown, from IQTREE tutorial</figcaption><p></p>
</figure>
</div>
<p><a href="http://www.iqtree.org/workshop/molevol2022">Source: Jeremy Brown</a></p>
<p>Importantly, different groups have gotten different answers even with 100% bootstrap support at the nodes. This highlights a problem that can occur when we have gene trees that conflict with the species tree. This can occur through incomplete lineage sorting (you will see ILS a lot in the literature; <a href="#fig-ils">Figure&nbsp;3</a>), or via hybridization, or other mechanisms.</p>
<div id="fig-ils" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="../../images/Incomplete-Lineage-Sorting.jpg" class="img-fluid figure-img"></p>
<p></p><figcaption class="figure-caption">Figure&nbsp;3: Incomplete Lineage Sorting. The species tree (thick gray history) in relation to a gene tree (thin colored history). New lineages arising from murations in the gene are represented by changes in color. Differences in the history of the gene versus the species can occur when the two alleles in the population prior to the split of Dmel are maintained through to the split of Dere and Dyak, leading to maintenance of an ancestral polymorphism and incomplete lineage sorting. The species tree is now incongruent with the gene tree (tree 2). The greater the diversity in the ancestral population and the shorter the time between speciation events, the more likely we are to see gene-tree species tree conflict.</figcaption><p></p>
</figure>
</div>
<p>Source: <span class="citation" data-cites="Pollard:2006">(<a href="#ref-Pollard:2006" role="doc-biblioref">Pollard et al. 2006</a>)</span></p>
</section>
<section id="the-data" class="level2">
<h2 class="anchored" data-anchor-id="the-data">The data</h2>
<p>The original genomic data comes from <span class="citation" data-cites="Chiari:2012">(<a href="#ref-Chiari:2012" role="doc-biblioref">Chiari et al. 2012</a>)</span>. We are analyzing a subset of 29 genes from the original 248 genes, as well as a subset of taxa (to allow faster runs), presented in Bui Minhʻs <a href="http://www.iqtree.org/workshop/molevol2022">tutorial</a>. We will use IQTREE2 to generate the phylogenetic data, and then use R to analyze some of the output. We are just going to do the basics, for more indepth coverage see <a href="http://www.iqtree.org/workshop/molevol2022">Minh Buiʻs tutorial</a>.</p>
</section>
<section id="species-trees" class="level2">
<h2 class="anchored" data-anchor-id="species-trees">Species Trees</h2>
<p>A shell script for thes IQTREE commands are in <code>iqtreerun.sh</code></p>
<section id="simplest-phylogeny" class="level3">
<h3 class="anchored" data-anchor-id="simplest-phylogeny">Simplest phylogeny</h3>
<p><a href="http://www.iqtree.org">IQTREE2</a> is the latest version of the IQTREE software. It runs on the command line, and each run generates a number of output files. To keep organized, it is a good idea to save your iqtree commands as a shell script, and to make input and output directories.</p>
<p>Mac users open your <code>Terminal</code>, Windows users open your <code>Git-Bash</code>, navigate to your <code>TurtleTutorial</code> folder.</p>
<p>The simple command below will find the best-fit model using ModelFinder, reconstruct the ML tree, and calculate branch supports using the ultrafast boostrap.</p>
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>Terminal</strong></pre>
</div>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="ex">iqtree2</span> <span class="at">-s</span> input/turtle.fa <span class="at">-B</span> 1000 <span class="at">-T</span> AUTO</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><strong>Run options explained:</strong></p>
<ul>
<li><code>-s input/turtle.fa</code> the path to the input alignment in <code>turtle.fa</code>.<br>
</li>
<li><code>-B 1000</code> 1000 replicates for the ultrafast bootstrap (Minh et al., 2013).<br>
</li>
<li><code>-T AUTO</code> auto select the best number of CPU cores to speed up the analysis.</li>
</ul>
<p><strong>Output files (in <code>out</code>):</strong></p>
<ul>
<li><code>turtle.fa.iqtree</code> the main report file that is self-readable. You should look at this file to see the computational results. It also contains a textual representation of the final tree.<br>
</li>
<li><code>turtle.fa.treefile</code> the ML tree in NEWICK format, which can be visualized in FigTree or any other tree viewer program.<br>
</li>
<li><code>turtle.fa.log</code> log file of the entire run (also printed on the screen).<br>
</li>
<li><code>turtle.fa.ckp.gz</code> checkpoint file used to resume an interrupted analysis.<br>
… and a few other files.</li>
</ul>
<div class="callout-tip callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Questions
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li>Look at the report file <code>turtle.fa.iqtree</code>, and the log file <code>turtle.fa.log</code>.</li>
<li>What is the best-fit model name? What do you know about this model? (see <a href="http://www.iqtree.org/doc/Substitution-Models">substitution models</a> available in IQ-TREE)</li>
<li>What are the AIC/AICc/BIC scores of this model and tree?</li>
<li>Visualise the tree <code>turtle.fa.treefile</code> in a tree viewer software like <a href="http://tree.bio.ed.ac.uk/software/figtree/">FigTree</a> (or read into R with <code>ape</code> and <code>plot()</code>).<br>
</li>
<li>What relationship among three trees does this tree support?</li>
<li>What is the ultrafast bootstrap support (%) for the relevant clade?</li>
<li>Does this tree agree with the published tree <span class="citation" data-cites="Chiari:2012">(<a href="#ref-Chiari:2012" role="doc-biblioref">Chiari et al. 2012</a>)</span>?</li>
</ul>
</div>
</div>
</section>
<section id="species-tree-partitioned-by-locus" class="level3">
<h3 class="anchored" data-anchor-id="species-tree-partitioned-by-locus">Species Tree partitioned by locus</h3>
<p>It is well known that some genes evolve faster than others (for example nuclear protein-coding genes vs.&nbsp;mitochondrial genes vs.&nbsp;non-coding genes). Partitioning by locus allows the rate of substitution for each locus to be scaled to different evolutionary rates for an improved fit of the evolutionary model to the observed DNA sequence data.</p>
<p>We add a partition file indicating where the loci are in the alighment:</p>
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>Terminal</strong></pre>
</div>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># infer the species tree with 1000 ultrafast bootstraps </span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="co"># and an edge-linked fully-partitioned model </span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="ex">iqtree2</span> <span class="at">-s</span> input/turtle.fa <span class="at">-p</span> input/turtle.nex <span class="at">--prefix</span> out/species <span class="at">-B</span> 1000 <span class="at">-nt</span> AUTO</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><strong>Run options explained:</strong></p>
<ul>
<li><code>-p input/turtle.nex</code> the partition file that delimits the partitions in the alighment to specify an edge-linked proportional partition model (Chernomor et al., 2016). There is one set of branch lengths for the phylogeny, but each partition can have proportionally shorter or longer tree length (= same as slower or faster evolutionary rates, respectively).</li>
</ul>
<div class="callout-tip callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Questions
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li>Look at the report file <code>turtle.fa.iqtree</code>, and the log file <code>turtle.fa.log</code>.</li>
<li>What are the AIC/AICc/BIC scores of this model and tree? How does it compare?</li>
<li>What relationship among three trees does this tree support?</li>
</ul>
</div>
</div>
</section>
<section id="gene-trees" class="level3">
<h3 class="anchored" data-anchor-id="gene-trees">Gene Trees</h3>
<p>We will now infer a separate phylogeny for each locus (= gene). It is really easy to compute gene trees in IQTREE. The same inputs are used as for the paritioned model, but we just tell IQTREE to estimate each gene tree separately with a different (<code>-S</code>) flag for the partition file:</p>
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>Terminal</strong></pre>
</div>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># infer single-locus trees  </span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="ex">iqtree2</span> <span class="at">-s</span> input/turtle.fa <span class="at">-S</span> input/turtle.nex <span class="at">-m</span> TEST <span class="at">--prefix</span> out/loci <span class="at">-B</span> 1000 </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><strong>Run options explained:</strong></p>
<ul>
<li><code>-S input/turtle.nex</code> to infer separate trees for every partition in <code>turtle.nex</code>. All output files are similar to a partition analysis, except that the tree <code>turtle.loci.treefile</code> now contains multiple trees - one for each gene.</li>
</ul>
</section>
</section>
<section id="concordance-factors" class="level2">
<h2 class="anchored" data-anchor-id="concordance-factors">Concordance factors</h2>
<p>Typcial phylogenetic inference methods essentially imply that gene trees and species tree are one and the same. However, it is well known that gene trees might be discordant (i.e., individual genes may have differet histories). We can actually check with large multilocus datasets (with many loci of sufficient length). Therefore, we now want to quantify the agreement between gene trees and species tree in a so-called concordance factor <span class="citation" data-cites="Minh:2020a">(<a href="#ref-Minh:2020a" role="doc-biblioref">Minh, Hahn, and Lanfear 2020</a>)</span>.</p>
<p>You can now compute gene concordance factor (gCF) and site concordance factor (sCF) for the tree inferred under the partition model:</p>
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>Terminal</strong></pre>
</div>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># locus modelfinder species tree vs. single-locus trees </span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="ex">iqtree2</span> <span class="at">-t</span> out/species.treefile <span class="at">--gcf</span> out/loci.treefile <span class="at">-s</span> input/turtle.fa <span class="at">--scf</span> 100 <span class="at">--prefix</span> out/concord </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><strong>Run options explained:</strong></p>
<ul>
<li><code>-t out/species.treefile</code> to specify the species tree. We use tree under the partitioned model here, but you can of course use the other tree.</li>
<li><code>--gcf out/loci.treefile</code> to specify a gene-trees file.</li>
<li><code>--scf 100</code> to draw 100 random quartets when computing sCF.</li>
</ul>
<p><strong>Output files (in <code>out</code>):</strong></p>
<ul>
<li><code>concord.cf.tree</code>: tree file where branches are annotated with bootstrap/gCF/sCF values.</li>
<li><code>concord.cf.stat</code>: a table file with various statistics for every branch of the tree.</li>
</ul>
<p>Similarly, you can compute gCF and sCF for the tree under different species trees, for example if you generate an unpartitioned model:</p>
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>Terminal</strong></pre>
</div>
<div class="sourceCode" id="cb5"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="ex">iqtree2</span> <span class="at">-t</span> out/unpartitioned.treefile <span class="at">--gcf</span> out/loci.treefile <span class="at">-s</span> input/turtle.fa <span class="at">--scf</span> 100 <span class="at">--prefix</span> out/concord.unpartitioned </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout-tip callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Tip
</div>
</div>
<div class="callout-body-container callout-body">
<p>If you need to redo any of these analyses, add the <code>-redo</code> flag at the end of the command.</p>
</div>
</div>
<div class="callout-note callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Definitions
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li>Gene concordance factor (gCF) is the percentage of decisive gene trees concordant with a particular branch of the species tree (0% &lt;= gCF(b) &lt;= 100%). gCF=0% means that branch b does not occur in any gene trees, whereas gCF=100% means that branch b occurs in every gene tree.<br>
</li>
<li>Site concordance factor (sCF) is the percentage of decisive (parsimony informative) alignment sites supporting a particular branch of the species tree (~33% &lt;= sCF(b) &lt;= 100%). sCF&lt;33% means that another discordant branch b’ is more supported, whereas sCF=100% means that branch b is supported by all sites.<br>
</li>
<li><strong>CAUTION</strong> when gCF ~ 0% or sCF &lt; 33%, even if boostrap supports are ~100%!</li>
<li><strong>GREAT</strong> when gCF and sCF &gt; 50% (i.e., branch is supported by a majority of genes and sites).</li>
</ul>
</div>
</div>
<div class="callout-tip callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Questions
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li>Visualise <code>concord.cf.tree.nex</code> in FigTree.</li>
<li>Explore gene concordance factor (gCF), gene discordance factors (gDF1, gDF2, gDFP), site concordance factor (sCF) and site discordance factors (sDF1, sDF2).</li>
<li>How do gCF and sCF values look compared with bootstrap supports?</li>
</ul>
</div>
</div>
</section>
</section>
<section id="further-analyses-of-iqtree-outputs" class="level1">
<h1>Further analyses of IQTREE outputs</h1>
<p>The code and analysis that follows was borrowed and adapted from Rob Lanfearʻs excellent blog on exploring gene-tree species-tree concordance <a href="http://www.robertlanfear.com/blog/files/concordance_factors.html" class="uri">http://www.robertlanfear.com/blog/files/concordance_factors.html</a>. This is an excellent example of how to integrate output from other analysis software with R code.</p>
<p>An R script for these analyses is in <code>concordance.R</code></p>
<section id="using-concordance-factors-to-understand-your-data" class="level2">
<h2 class="anchored" data-anchor-id="using-concordance-factors-to-understand-your-data">Using concordance factors to understand your data</h2>
<p>The first step is just to look at the concordance factors on the tree. To do this, load the output tree (concord.cf.tree) in any tree viewer such as Figtree or you can use R. In this tree, each branch label shows bootstrap / gCF / sCF.</p>
<p>We can also plot the trees using <code>ggtree</code> (but for a first look around I usually use FigTree).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">require</span>(viridis)</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="fu">require</span>(GGally)</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a><span class="fu">require</span>(entropy)</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a><span class="fu">require</span>(ggtree)</span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a><span class="fu">require</span>(treeio)</span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a><span class="fu">require</span>(tidytree)</span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a><span class="fu">require</span>(ggplot2)</span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a><span class="fu">require</span>(dplyr)</span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a><span class="fu">require</span>(ggrepel)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co"># read the data</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>d <span class="ot">=</span> <span class="fu">read.delim</span>(<span class="st">"out/concord.cf.stat"</span>, <span class="at">header =</span> T, <span class="at">comment.char=</span><span class="st">'#'</span>)</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>               </span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a><span class="co"># rename the bootstrap and branchlength columns</span></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(d)[<span class="fu">names</span>(d)<span class="sc">==</span><span class="st">"Label"</span>] <span class="ot">=</span> <span class="st">"bootstrap"</span></span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(d)[<span class="fu">names</span>(d)<span class="sc">==</span><span class="st">"Length"</span>] <span class="ot">=</span> <span class="st">"branchlength"</span></span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a><span class="co"># plot the tree</span></span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a>tree <span class="ot">&lt;-</span> <span class="fu">read.iqtree</span>(<span class="st">"out/concord.cf.tree"</span>)  <span class="co"># read in tree file</span></span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a>tib <span class="ot">&lt;-</span> <span class="fu">as_tibble</span>(tree)</span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true" tabindex="-1"></a>d<span class="sc">$</span>node <span class="ot">&lt;-</span> d<span class="sc">$</span>ID <span class="sc">+</span> <span class="dv">1</span>  <span class="co"># need to +1 to make IQTREE node numbers = ggtree </span></span>
<span id="cb7-12"><a href="#cb7-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-13"><a href="#cb7-13" aria-hidden="true" tabindex="-1"></a><span class="co"># merge the tree and data </span></span>
<span id="cb7-14"><a href="#cb7-14" aria-hidden="true" tabindex="-1"></a>td <span class="ot">&lt;-</span> <span class="fu">full_join</span>(tib, d, <span class="at">by =</span> <span class="st">"node"</span>)  <span class="co"># combine the tree (tib) and data (d)</span></span>
<span id="cb7-15"><a href="#cb7-15" aria-hidden="true" tabindex="-1"></a>td <span class="ot">&lt;-</span> <span class="fu">as.treedata</span>(td)       <span class="co"># coerce to treedata format (td)</span></span>
<span id="cb7-16"><a href="#cb7-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-17"><a href="#cb7-17" aria-hidden="true" tabindex="-1"></a><span class="fu">ggtree</span>(td) <span class="sc">+</span>          <span class="co"># plot tree</span></span>
<span id="cb7-18"><a href="#cb7-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position=</span><span class="st">"right"</span>) <span class="sc">+</span></span>
<span id="cb7-19"><a href="#cb7-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_tiplab</span>() <span class="sc">+</span>                         <span class="co"># add tip labels</span></span>
<span id="cb7-20"><a href="#cb7-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_label2</span>(<span class="fu">aes</span>(<span class="at">label=</span>label, <span class="at">subset=</span><span class="sc">!</span>isTip), <span class="at">color=</span><span class="st">"black"</span>)  <span class="co"># add node labels </span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="index_files/figure-html/unnamed-chunk-2-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>The turtle clade contains <strong>Chelonoidis</strong> to <strong>Phrynops</strong>. Which clade do turtles group with? This part of the tree immediately illustrates the most important point: bootstraps and concordance factors are giving you very different information about each branch in the tree. Just take a look at how different the three numbers can be!</p>
</section>
</section>
<section id="how-concordance-factors-relate-to-each-other-and-to-bootstraps" class="level1">
<h1>How concordance factors relate to each other and to bootstraps</h1>
<p>More generally, we can look at the links between bootstrap, gCF, and sCF across all nodes of the tree. This is simple to do in R, because IQ-TREE outputs a tab-delimited file that’s easy to read called concord.cf.stat. This file has one row per branch in the tree, and gives a lot of details on the statistics for each branch.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># plot the values</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(d, <span class="fu">aes</span>(<span class="at">x =</span> gCF, <span class="at">y =</span> sCF)) <span class="sc">+</span> </span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">colour =</span> bootstrap)) <span class="sc">+</span> </span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_colour_viridis</span>(<span class="at">direction =</span> <span class="sc">-</span><span class="dv">1</span>) <span class="sc">+</span> </span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">xlim</span>(<span class="dv">0</span>, <span class="dv">100</span>) <span class="sc">+</span></span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ylim</span>(<span class="dv">0</span>, <span class="dv">100</span>) <span class="sc">+</span></span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_abline</span>(<span class="at">slope =</span> <span class="dv">1</span>, <span class="at">intercept =</span> <span class="dv">0</span>, <span class="at">linetype =</span> <span class="st">"dashed"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="index_files/figure-html/unnamed-chunk-3-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>This plot shows a few important things. First, <em>low bootstrap values (bright colours) coincide with the lowest gCF and sCF values, as expected</em>. You can only get a low bootstrap value when the there’s very limited information on that branch in the alignment. And if there’s very little information, there will be very few sites (and therefore genes) that can support a branch.</p>
<p><strong>However, it is possible to have low gCF and sCF even with high bootstrap values.</strong> It turns out that <em>bootstrap values max out at 100% pretty quickly</em>, which tends to happen with very large datasets such as provided by Next-Gen sequencing.</p>
<p>Finally, <strong>sCF values have a minimum of ~30%</strong>, but <strong>gCF values can go all the way to 0%</strong>. This means that an sCF value ~30% is as low as you can possibly get. The range of possible values are simply a reflection of how these statistics are calculated. sCF values are calculated by comparing the three possible resolutions of quartet around a node, so when the data are completely equivocal about these resolutions, we expect an sCF value of 1 out of 3 or 33%. gCF values, on the other hand, are calculated from full gene trees, such that there are many more than 3 possible resolutions around a node, and the gCF value can be as low as 0% if no single gene tree contains a branch that’s present in the reference tree. This can happen when the genes are too short to contain informative sites (and happens more often deeper in the tree), or noisy data – a combination of biology and stochastic error that leads to a lot of gene-tree discordance.</p>
<p>Sometimes we want to label points. <code>ggrepel()</code> is very helpful for this:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co"># label the points</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(d, <span class="fu">aes</span>(<span class="at">x =</span> gCF, <span class="at">y =</span> sCF, <span class="at">label =</span> ID)) <span class="sc">+</span> </span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">colour =</span> bootstrap)) <span class="sc">+</span> </span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_colour_viridis</span>(<span class="at">direction =</span> <span class="sc">-</span><span class="dv">1</span>) <span class="sc">+</span> </span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">xlim</span>(<span class="dv">0</span>, <span class="dv">100</span>) <span class="sc">+</span></span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ylim</span>(<span class="dv">0</span>, <span class="dv">100</span>) <span class="sc">+</span></span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_abline</span>(<span class="at">slope =</span> <span class="dv">1</span>, <span class="at">intercept =</span> <span class="dv">0</span>, <span class="at">linetype =</span> <span class="st">"dashed"</span>) <span class="sc">+</span></span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_text_repel</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="index_files/figure-html/unnamed-chunk-4-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="digging-deeper-using-discordance-factors" class="level1">
<h1>Digging deeper using discordance factors</h1>
<p>First we need to find the ID of the problematic node:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co"># show branches of interest</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>d[(d<span class="sc">$</span>ID<span class="sc">==</span><span class="dv">20</span><span class="sc">|</span>d<span class="sc">$</span>ID<span class="sc">==</span><span class="dv">22</span>),]  <span class="co"># these have lower bootstrap support than the others and have low gCF or sCF</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  ID   gCF gCF_N gDF1 gDF1_N  gDF2 gDF2_N  gDFP gDFP_N gN   sCF sCF_N  sDF1
4 20 75.00     3 0.00      0 25.00      1  0.00      0  4 38.24 26.35 28.02
6 22 48.28    14 3.45      1 17.24      5 31.03      9 29 36.13 57.59 22.17
  sDF1_N  sDF2 sDF2_N     sN bootstrap branchlength node
4  19.59 33.75  23.71  69.65        51   0.00355984   21
6  37.78 39.70  72.90 168.27        81   0.01865970   23</code></pre>
</div>
</div>
<p>We can see that the problematic nodes have very short branch lengths, which is consistent with possible incomplete lineage sorting.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="co"># plot the tree with colors to highlight support</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a><span class="fu">ggtree</span>(td, <span class="fu">aes</span>(<span class="at">color=</span>SH_aLRT), <span class="at">size=</span><span class="dv">2</span>) <span class="sc">+</span>          <span class="co"># plot tree</span></span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position=</span><span class="st">"right"</span>) <span class="sc">+</span></span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_tiplab</span>(<span class="at">size=</span><span class="dv">5</span>) <span class="sc">+</span>                         <span class="co"># add tip labels</span></span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_label2</span>(<span class="fu">aes</span>(<span class="at">label=</span>label, <span class="at">subset=</span><span class="sc">!</span>isTip), <span class="at">color=</span><span class="st">"black"</span>) <span class="sc">+</span> <span class="co"># add node labels </span></span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_label2</span>(<span class="fu">aes</span>(<span class="at">label=</span>label, <span class="at">subset=</span> (<span class="sc">!</span>isTip <span class="sc">&amp;</span> (node<span class="sc">==</span><span class="dv">22</span>))), <span class="at">color=</span><span class="st">"black"</span>, <span class="at">fill=</span><span class="st">"yellow"</span>) <span class="sc">+</span>                  <span class="co"># highlight low bootstrap support </span></span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_continuous</span>(<span class="at">low=</span><span class="st">"yellow"</span>, <span class="at">high=</span><span class="st">"purple"</span>) <span class="co"># color the branches by bootstrap</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="index_files/figure-html/unnamed-chunk-6-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggtree</span>(td) <span class="sc">+</span>                                      <span class="co"># plot tree</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_tiplab</span>() <span class="sc">+</span>                               <span class="co"># plot tip labels</span></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_label2</span>(<span class="fu">aes</span>(<span class="at">label=</span>node, <span class="at">subset=</span> <span class="sc">!</span>isTip))  <span class="co"># plot node numbers </span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="index_files/figure-html/unnamed-chunk-6-2.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="using-concordance-factors-to-test-the-assumptions-of-an-ils-model" class="level1">
<h1>Using concordance factors to test the assumptions of an ILS model</h1>
<p>If the discordance among gene trees and sites come from <em>incomplete lineage sorting</em>, different genes should randomly support one or the other of the discordant topologies. Thus, we can make a simple and testable prediction: that the number of gene trees or sites supporting the two discordant topologies should be roughly equal. Both of these ideas have been around for some time (for genes [link to Huson et al &amp; Steel 2005 Recomb] and for sites [link to Greene et al 2010]), and the IQ-TREE output lets us test them very easily.</p>
<p>The basic idea is that we count up the genes or sites supporting the two discordant topologies, and use a chi-square test to see if they’re significantly different. This requires a number of assumptions to hold (see the papers linked above). A simple way to calculate the probability that the data can reject <strong>equal frequencies</strong> of discordant toplogogies (EFp) for genes (gEFp) and for sites (sEFp). In this case failure to reject the hypothesis of equal frequencies is compatible with ILS.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="co"># first we use a slightly modified chisq function</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a><span class="co"># which behaves nicely when you feed it zeros</span></span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>chisq <span class="ot">=</span> <span class="cf">function</span>(DF1, DF2, N){</span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">tryCatch</span>({</span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>        <span class="co"># converts percentages to counts, runs chisq, gets pvalue</span></span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>        <span class="fu">chisq.test</span>(<span class="fu">c</span>(<span class="fu">round</span>(DF1<span class="sc">*</span>N)<span class="sc">/</span><span class="dv">100</span>, <span class="fu">round</span>(DF2<span class="sc">*</span>N)<span class="sc">/</span><span class="dv">100</span>))<span class="sc">$</span>p.value</span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a>    },</span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">error =</span> <span class="cf">function</span>(err) {</span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a>        <span class="co"># errors come if you give chisq two zeros</span></span>
<span id="cb14-10"><a href="#cb14-10" aria-hidden="true" tabindex="-1"></a>        <span class="co"># but here we're sure that there's no difference</span></span>
<span id="cb14-11"><a href="#cb14-11" aria-hidden="true" tabindex="-1"></a>        <span class="fu">return</span>(<span class="fl">1.0</span>)</span>
<span id="cb14-12"><a href="#cb14-12" aria-hidden="true" tabindex="-1"></a>    })</span>
<span id="cb14-13"><a href="#cb14-13" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb14-14"><a href="#cb14-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-15"><a href="#cb14-15" aria-hidden="true" tabindex="-1"></a>e <span class="ot">=</span> d <span class="sc">%&gt;%</span> </span>
<span id="cb14-16"><a href="#cb14-16" aria-hidden="true" tabindex="-1"></a>    <span class="fu">group_by</span>(ID) <span class="sc">%&gt;%</span></span>
<span id="cb14-17"><a href="#cb14-17" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">gEF_p =</span> <span class="fu">chisq</span>(gDF1, gDF2, gN)) <span class="sc">%&gt;%</span></span>
<span id="cb14-18"><a href="#cb14-18" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">sEF_p =</span> <span class="fu">chisq</span>(sDF1, sDF2, sN))</span>
<span id="cb14-19"><a href="#cb14-19" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb14-20"><a href="#cb14-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-21"><a href="#cb14-21" aria-hidden="true" tabindex="-1"></a><span class="fu">subset</span>(<span class="fu">data.frame</span>(e), (gEF_p <span class="sc">&lt;</span> <span class="fl">0.05</span> <span class="sc">|</span> sEF_p <span class="sc">&lt;</span> <span class="fl">0.05</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  ID   gCF gCF_N gDF1 gDF1_N  gDF2 gDF2_N  gDFP gDFP_N gN   sCF sCF_N  sDF1
6 22 48.28    14 3.45      1 17.24      5 31.03      9 29 36.13 57.59 22.17
  sDF1_N sDF2 sDF2_N     sN bootstrap branchlength node     gEF_p       sEF_p
6  37.78 39.7   72.9 168.27        81    0.0186597   23 0.1024704 0.003849915</code></pre>
</div>
</div>
</section>
<section id="internode-certainty" class="level1">
<h1>Internode certainty</h1>
<p>The last measure to quantify evidence for each branch is the <strong>internode certainty</strong> <span class="citation" data-cites="Salichos:2014">(<a href="#ref-Salichos:2014" role="doc-biblioref">Salichos, Stamatakis, and Rokas 2014</a>)</span>.</p>
<p>the conflict along a given branch is</p>
<p>Internode certainty quantifies the degree of certainty for each branch (individual internode). It finds the two most common conflicting bipartitions, and measures the log magnitude of their difference.</p>
<p>We determine concordance in a clever way using set theory. We can think of each branch as splitting the tree into two subsets of taxa (bipartitions). What is so clever is that we donʻt have to compare the toplogies of the subtrees, just the taxa included. The task then comes down to checking for compatible intersections of the bipartitions. This makes it very easy (=fast) computationally. Hereʻs a figure from the paper by <a href="#fig-internode-conflict">Figure&nbsp;4</a></p>
<div id="fig-internode-conflict" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="../../images/internode-conflict.png" class="img-fluid figure-img"></p>
<p></p><figcaption class="figure-caption">Figure&nbsp;4: Compatible and conflicting bipartitions. Bipartition A = {a, b, c, d, e | f, g, h, i, j} is composed of the partitions A1 = {{a, b, c, d, e}} and A2 = {f, g, h, i, j}, where a, b, c, d, e, f, g, h, i, and j are taxa. Bipartition B = {a, b, c | d, e, f, g, h, i, j} is composed of the partitions B1 = {a, b, c} and B2 = {d, e, f, g, h, i, j}, and bipartition C = {a, b, c, d, g | e, f, h, i, j} is composed of the partitions C1 = {a, b, c, d, g} and C2 = {e, f, h, i, j}.</figcaption><p></p>
</figure>
</div>
<p><span class="citation" data-cites="Salichos:2014">(Source: <a href="#ref-Salichos:2014" role="doc-biblioref">Salichos, Stamatakis, and Rokas 2014</a>)</span></p>
<p>We check the compatibility of the biparitions by checking the intersections of their each possible set of bipartition pairs.</p>
<div class="callout-note callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
The definition of compatibility
</div>
</div>
<div class="callout-body-container callout-body">
<p>Two bipartitions <span class="math inline">\(A=A_1 | A_2\)</span> and <span class="math inline">\(B=B_1 | B_2\)</span> from the same taxon set are <strong>compatible</strong> <em>if and only if</em> at least one of the intersections of the four bipartition pairs (<span class="math inline">\(A_1 \cap B_1\)</span>, <span class="math inline">\(A_1 \cap B_2\)</span>, <span class="math inline">\(A_2 \cap B_1\)</span>, <span class="math inline">\(A_2 \cap B_2\)</span>) is empty (see citations in <span class="citation" data-cites="Salichos:2014">Salichos, Stamatakis, and Rokas (<a href="#ref-Salichos:2014" role="doc-biblioref">2014</a>)</span>).</p>
</div>
</div>
<p>Bipartitions A and B are compatible because one of the intersections of their bipartition pairs (<span class="math inline">\(A_2 \cap B_1\)</span>) is empty.</p>
<p><span class="math display">\[
A_2 \cap B_1 =  \{f, g, h, i, j\} \cap \{a, b, c\} = \emptyset
\]</span></p>
<p>Bipartitions B and C are also compatible (<span class="math inline">\(B_1 \cap C_2\)</span> is empty).</p>
<p><span class="math display">\[
B_1 \cap C_2 =  \{a, b, c\} \cap \{e, f, h, i, j\} = \emptyset
\]</span></p>
<p>In contrast, bipartition C conflicts with A (or is incompatible) because none of the four intersections are empty:</p>
<p><span class="math display">\[
\begin{aligned}
A_1 \cap C_1 &amp;=  \{{a, b, c, d, e}\} \cap \{a, b, c, d, g\} = \{a, b, c, d\}  \\
A_1 \cap C_2 &amp;=  \{{a, b, c, d, e}\} \cap \{e, f, h, i, j\} = \{e\}  \\
A_2 \cap C_1 &amp;=  \{{f, g, h, i, j}\} \cap \{a, b, c, d, g\} = \{a, b, c, d\}  \\  
A_2 \cap C_2 &amp;=  \{{f, g, h, i, j}\} \cap \{e, f, h, i, j\} = \{e\}  
\end{aligned}
\]</span></p>
<p>For a given branch, Internode Certainty (IC) values reflect the degree of conflict between the bipartition defining the branch (say X) and the most frequent conflicting biparition (say Y), which we can get by counting the number of concordant genes or sites for each bipartition. IC is then defined by:</p>
<p><span class="math display">\[
\begin{aligned}
IC &amp;= log_2(n) + \sum_{n=1}^2 P(X_n)log_2[P(X_n)] \\
   &amp;= 1 + P(X)log_2[P(X)] + P(Y)log_2[P(Y)],
\end{aligned}
\]</span> where the frequencies of the bipartitons are:</p>
<p><span class="math display">\[
\begin{aligned}
P(X) = X/(X + Y), \\  
P(Y) = Y/(X + Y)
\end{aligned}
\]</span> and <span class="math inline">\(P(X) + P(Y) = 1.\)</span></p>
<p>Therefore, IC values at or close to 1 indicate support for the given branch (i.e., absence of conflict), and IC values at or close to 0 indicate maximum conflict (i.e., equal support for both bipartitions).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate internode certainty</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>IC <span class="ot">=</span> <span class="cf">function</span>(CF, DF1, DF2, N){</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a>    <span class="co"># convert to counts</span></span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a>    X <span class="ot">=</span> CF <span class="sc">*</span> N <span class="sc">/</span> <span class="dv">100</span></span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a>    Y <span class="ot">=</span> <span class="fu">max</span>(DF1, DF2) <span class="sc">*</span> N <span class="sc">/</span> <span class="dv">100</span></span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb16-9"><a href="#cb16-9" aria-hidden="true" tabindex="-1"></a>    pX <span class="ot">=</span> X<span class="sc">/</span>(X<span class="sc">+</span>Y)</span>
<span id="cb16-10"><a href="#cb16-10" aria-hidden="true" tabindex="-1"></a>    pY <span class="ot">=</span> Y<span class="sc">/</span>(X<span class="sc">+</span>Y)</span>
<span id="cb16-11"><a href="#cb16-11" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb16-12"><a href="#cb16-12" aria-hidden="true" tabindex="-1"></a>    IC <span class="ot">=</span> <span class="dv">1</span> <span class="sc">+</span> pX <span class="sc">*</span> <span class="fu">log2</span>(pX) <span class="sc">+</span></span>
<span id="cb16-13"><a href="#cb16-13" aria-hidden="true" tabindex="-1"></a>             pY <span class="sc">*</span> <span class="fu">log2</span>(pY)</span>
<span id="cb16-14"><a href="#cb16-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-15"><a href="#cb16-15" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>(IC)</span>
<span id="cb16-16"><a href="#cb16-16" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb16-17"><a href="#cb16-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-18"><a href="#cb16-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-19"><a href="#cb16-19" aria-hidden="true" tabindex="-1"></a>e <span class="ot">=</span> e <span class="sc">%&gt;%</span> </span>
<span id="cb16-20"><a href="#cb16-20" aria-hidden="true" tabindex="-1"></a>    <span class="fu">group_by</span>(ID) <span class="sc">%&gt;%</span></span>
<span id="cb16-21"><a href="#cb16-21" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">gIC =</span> <span class="fu">IC</span>(gCF, gDF1, gDF2, gN)) <span class="sc">%&gt;%</span></span>
<span id="cb16-22"><a href="#cb16-22" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">sIC =</span> <span class="fu">IC</span>(sCF, sDF1, sDF2, sN))</span>
<span id="cb16-23"><a href="#cb16-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-24"><a href="#cb16-24" aria-hidden="true" tabindex="-1"></a><span class="co"># plot it</span></span>
<span id="cb16-25"><a href="#cb16-25" aria-hidden="true" tabindex="-1"></a><span class="fu">ggpairs</span>(e, <span class="at">columns =</span> <span class="fu">c</span>(<span class="st">"gCF"</span>, <span class="st">"sCF"</span>, <span class="st">"bootstrap"</span>, <span class="st">"gEF_p"</span>, <span class="st">"sEF_p"</span>, <span class="st">"gIC"</span>, <span class="st">"sIC"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="index_files/figure-html/unnamed-chunk-8-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="callout-tip callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Questions
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li>Which nodes have IC near zero?</li>
<li>How does it correspond to gCF, sCF, and bootstrap support?</li>
</ul>
</div>
</div>
<p>See Rob Lanfearʻs blog for detailed explanations of the following calculations <a href="http://www.robertlanfear.com/blog/files/concordance_factors.html" class="uri">http://www.robertlanfear.com/blog/files/concordance_factors.html</a>. This is an excellent example of how to integrate output from other analysis software with R code.</p>
</section>
<section id="iqtree-documentation" class="level1">
<h1>IQTREE documentation</h1>
<ul>
<li><p>[IQTREE quickstart]](http://www.iqtree.org/doc/Quickstart)</p></li>
<li><p>IQTREE documentation <a href="http://www.iqtree.org/doc/" class="uri">http://www.iqtree.org/doc/</a></p></li>
<li><p>Command Line Reference (when you get more familiar with it) <a href="http://www.iqtree.org/doc/Command-Reference" class="uri">http://www.iqtree.org/doc/Command-Reference</a></p></li>
</ul>



</section>

<div id="quarto-appendix" class="default"><section class="quarto-appendix-contents" role="doc-bibliography"><h2 class="anchored quarto-appendix-heading">References</h2><div id="refs" class="references csl-bib-body hanging-indent" role="doc-bibliography">
<div id="ref-Chiari:2012" class="csl-entry" role="doc-biblioentry">
Chiari, Ylenia, Vincent Cahais, Nicolas Galtier, and Frédéric Delsuc. 2012. <span>“Phylogenomic Analyses Support the Position of Turtles as the Sister Group of Birds and Crocodiles (Archosauria).”</span> <em>BMC Biology</em> 10 (1): 65. <a href="https://doi.org/10.1186/1741-7007-10-65">https://doi.org/10.1186/1741-7007-10-65</a>.
</div>
<div id="ref-Kalyaanamoorthy:2017" class="csl-entry" role="doc-biblioentry">
Kalyaanamoorthy, Subha, Bui Quang Minh, Thomas K F Wong, Arndt von Haeseler, and Lars S Jermiin. 2017. <span>“ModelFinder: Fast Model Selection for Accurate Phylogenetic Estimates.”</span> <em>Nature Methods</em> 14 (6): 587–89. <a href="https://doi.org/10.1038/nmeth.4285">https://doi.org/10.1038/nmeth.4285</a>.
</div>
<div id="ref-Lanfear:2016" class="csl-entry" role="doc-biblioentry">
Lanfear, Robert, Paul B. Frandsen, April M. Wright, Tereza Senfeld, and Brett Calcott. 2016. <span>“<span class="nocase">PartitionFinder 2: New Methods for Selecting Partitioned Models of Evolution for Molecular and Morphological Phylogenetic Analyses</span>.”</span> <em>Molecular Biology and Evolution</em> 34 (3): 772–73. <a href="https://doi.org/10.1093/molbev/msw260">https://doi.org/10.1093/molbev/msw260</a>.
</div>
<div id="ref-Minh:2020a" class="csl-entry" role="doc-biblioentry">
Minh, Bui Quang, Matthew W Hahn, and Robert Lanfear. 2020. <span>“<span class="nocase">New Methods to Calculate Concordance Factors for Phylogenomic Datasets</span>.”</span> <em>Molecular Biology and Evolution</em> 37 (9): 2727–33. <a href="https://doi.org/10.1093/molbev/msaa106">https://doi.org/10.1093/molbev/msaa106</a>.
</div>
<div id="ref-Minh:2020" class="csl-entry" role="doc-biblioentry">
Minh, Bui Quang, Heiko A Schmidt, Olga Chernomor, Dominik Schrempf, Michael D Woodhams, Arndt von Haeseler, and Robert Lanfear. 2020. <span>“<span class="nocase">IQ-TREE 2: New Models and Efficient Methods for Phylogenetic Inference in the Genomic Era</span>.”</span> <em>Molecular Biology and Evolution</em> 37 (5): 1530–34. <a href="https://doi.org/10.1093/molbev/msaa015">https://doi.org/10.1093/molbev/msaa015</a>.
</div>
<div id="ref-Pollard:2006" class="csl-entry" role="doc-biblioentry">
Pollard, Daniel A, Venky N Iyer, Alan M Moses, and Michael B Eisen. 2006. <span>“Widespread Discordance of Gene Trees with Species Tree in Drosophila: Evidence for Incomplete Lineage Sorting.”</span> Doi:10.1371/journal.pgen.0020173. <em>PLOS Genetics</em> 2 (10): e173–. <a href="https://doi.org/10.1371/journal.pgen.0020173">https://doi.org/10.1371/journal.pgen.0020173</a>.
</div>
<div id="ref-Salichos:2014" class="csl-entry" role="doc-biblioentry">
Salichos, Leonidas, Alexandros Stamatakis, and Antonis Rokas. 2014. <span>“<span class="nocase">Novel Information Theory-Based Measures for Quantifying Incongruence among Phylogenetic Trees</span>.”</span> <em>Molecular Biology and Evolution</em> 31 (5): 1261–71. <a href="https://doi.org/10.1093/molbev/msu061">https://doi.org/10.1093/molbev/msu061</a>.
</div>
</div></section></div></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>